{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Using `memento` to analyze Interferon-B response in monocytes\n",
    "\n",
    "To install `memento` in the pre-release version (for Ye Lab members), install it directly from github by running:\n",
    "\n",
    "```pip install git+https://github.com/yelabucsf/scrna-parameter-estimation.git@release-v0.0```\n",
    "\n",
    "This requires that you have access to the Ye Lab organization. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/data/home/anaconda3/envs/single_cell/lib/python3.7/site-packages/scanpy/api/__init__.py:6: FutureWarning: \n",
      "\n",
      "In a future version of Scanpy, `scanpy.api` will be removed.\n",
      "Simply use `import scanpy as sc` and `import scanpy.external as sce` instead.\n",
      "\n",
      "  FutureWarning,\n"
     ]
    }
   ],
   "source": [
    "import sys\n",
    "sys.path.append('/data/home/Github/scrna-parameter-estimation/dist/memento-0.0.1-py3.7.egg')\n",
    "import memento"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import memento"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "fig_path = '/data/home/Github/scrna-parameter-estimation/figures/fig4/'\n",
    "data_path = '/data/parameter_estimation/'"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read IFN data and filter for monocytes\n",
    "\n",
    "For `memento`, we need the raw count matrix. Preferrably, feed the one with all genes so that we can choose what genes to look at. \n",
    "\n",
    "One of the columns in `adata.obs` should be the discrete groups to compare mean, variability, and co-variability across. In this case, it's called `stim`. \n",
    "\n",
    "The column containing the covariate that you want p-values for should either:\n",
    "- Be binary (aka the column only contains two unique values, such as 'A' and 'B'. Here, the values are either 'stim' or 'ctrl'.\n",
    "- Be numeric (aka the column contains -1, 0, -1 for each genotype value). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 5341 × 35635 \n",
      "    obs: 'tsne1', 'tsne2', 'ind', 'stim', 'cluster', 'cell', 'multiplets', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'total_counts_mt', 'log1p_total_counts_mt', 'pct_counts_mt', 'total_counts_hb', 'log1p_total_counts_hb', 'pct_counts_hb', 'cell_type'\n",
      "    var: 'gene_ids', 'mt', 'hb', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts'\n",
      "    uns: 'cell_type_colors'\n",
      "    obsm: 'X_tsne'\n"
     ]
    }
   ],
   "source": [
    "adata = sc.read(data_path + 'interferon_filtered.h5ad')\n",
    "adata = adata[adata.obs.cell == 'CD14+ Monocytes'].copy()\n",
    "print(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ind</th>\n",
       "      <th>stim</th>\n",
       "      <th>cell</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>index</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>TATCTGACAAAAGC-1</th>\n",
       "      <td>1256</td>\n",
       "      <td>ctrl</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ATCGCAGATTCTAC-1</th>\n",
       "      <td>1488</td>\n",
       "      <td>ctrl</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTACGAACTAACCG-1</th>\n",
       "      <td>1015</td>\n",
       "      <td>ctrl</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CCATAGGACAGATC-1</th>\n",
       "      <td>1244</td>\n",
       "      <td>stim</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GTGACCCTAGAATG-1</th>\n",
       "      <td>1244</td>\n",
       "      <td>ctrl</td>\n",
       "      <td>CD14+ Monocytes</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                   ind  stim             cell\n",
       "index                                        \n",
       "TATCTGACAAAAGC-1  1256  ctrl  CD14+ Monocytes\n",
       "ATCGCAGATTCTAC-1  1488  ctrl  CD14+ Monocytes\n",
       "GTACGAACTAACCG-1  1015  ctrl  CD14+ Monocytes\n",
       "CCATAGGACAGATC-1  1244  stim  CD14+ Monocytes\n",
       "GTGACCCTAGAATG-1  1244  ctrl  CD14+ Monocytes"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs[['ind', 'stim', 'cell']].sample(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Create groups for hypothesis testing and compute 1D parameters\n",
    "\n",
    "`memento` creates groups of cells based on anything that should be considered a reasonable group; here, we just divide the cells into `stim` and `ctrl`. But we can easily further divide the cells into individuals by adding the `ind` column to the `label_columns` argument when calling `create_groups`.\n",
    "\n",
    "`q` is the rough estimate of the overall UMI efficiency across both sampling and sequencing. If `s` is the sequencing saturation, multiply `s` by 0.07 for 10X v1, 0.15 for v2, and 0.25 for v3. \n",
    "\n",
    "By default, `memento` will consider all genes whose expression is high enough to calculate an accurate variance. If you wish to include less genes, increase `filter_mean_thresh`."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "memento.create_groups(adata, label_columns=['stim'], inplace=True, q=0.07)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "memento.compute_1d_moments(\n",
    "    adata, \n",
    "    inplace=True, \n",
    "    filter_mean_thresh=0.07, # minimum raw mean of each gene within a group for the gene to be considered \n",
    "    min_perc_group=.9) # percentage of groups that satisfy the condition for a gene to be considered. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform 1D hypothesis testing\n",
    "\n",
    "`formula_like` determines the linear model that is used for hypothesis testing, while `cov_column` is used to pick out the variable that you actually want p-values for. \n",
    "\n",
    "`num_cpus` controls how many CPUs to parallelize this operation for. In general, I recommend using 3-6 CPUs for reasonable peformance on any of the AWS machines that we have access to (I'm currently using a c5.2xlarge instance (8 vCPUs). "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=6)]: Using backend LokyBackend with 6 concurrent workers.\n",
      "[Parallel(n_jobs=6)]: Done  64 tasks      | elapsed:    2.9s\n",
      "[Parallel(n_jobs=6)]: Done 364 tasks      | elapsed:   15.1s\n",
      "[Parallel(n_jobs=6)]: Done 864 tasks      | elapsed:   36.7s\n",
      "[Parallel(n_jobs=6)]: Done 1564 tasks      | elapsed:  1.1min\n",
      "[Parallel(n_jobs=6)]: Done 1866 out of 1877 | elapsed:  1.3min remaining:    0.5s\n",
      "[Parallel(n_jobs=6)]: Done 1877 out of 1877 | elapsed:  1.4min finished\n"
     ]
    }
   ],
   "source": [
    "memento.ht_1d_moments(\n",
    "    adata, \n",
    "    formula_like='1 + stim',\n",
    "    cov_column='stim', \n",
    "    num_boot=5000, \n",
    "    verbose=1,\n",
    "    num_cpus=6)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_1d = memento.get_1d_ht_result(adata)\n",
    "result_1d['de_fdr'] = memento.util._fdrcorrect(result_1d['de_pval'])\n",
    "result_1d['dv_fdr'] = memento.util._fdrcorrect(result_1d['dv_pval'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>de_coef</th>\n",
       "      <th>de_pval</th>\n",
       "      <th>dv_coef</th>\n",
       "      <th>dv_pval</th>\n",
       "      <th>de_fdr</th>\n",
       "      <th>dv_fdr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>357</th>\n",
       "      <td>RPL15</td>\n",
       "      <td>-0.830076</td>\n",
       "      <td>4.407524e-77</td>\n",
       "      <td>0.531724</td>\n",
       "      <td>0.000182</td>\n",
       "      <td>8.272923e-74</td>\n",
       "      <td>0.004828</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1158</th>\n",
       "      <td>GAPDH</td>\n",
       "      <td>-0.648887</td>\n",
       "      <td>9.882150e-67</td>\n",
       "      <td>0.033378</td>\n",
       "      <td>0.003000</td>\n",
       "      <td>9.274397e-64</td>\n",
       "      <td>0.047319</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>49</th>\n",
       "      <td>MARCKSL1</td>\n",
       "      <td>-1.277120</td>\n",
       "      <td>7.097472e-65</td>\n",
       "      <td>0.478701</td>\n",
       "      <td>0.053200</td>\n",
       "      <td>4.440651e-62</td>\n",
       "      <td>0.324209</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>856</th>\n",
       "      <td>RPL10</td>\n",
       "      <td>-0.666116</td>\n",
       "      <td>9.276986e-61</td>\n",
       "      <td>0.238853</td>\n",
       "      <td>0.599800</td>\n",
       "      <td>4.353226e-58</td>\n",
       "      <td>0.889168</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1105</th>\n",
       "      <td>FAU</td>\n",
       "      <td>-0.379072</td>\n",
       "      <td>2.031012e-54</td>\n",
       "      <td>0.525444</td>\n",
       "      <td>0.010200</td>\n",
       "      <td>7.624420e-52</td>\n",
       "      <td>0.117456</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1238</th>\n",
       "      <td>RPL6</td>\n",
       "      <td>-1.343297</td>\n",
       "      <td>6.179513e-54</td>\n",
       "      <td>0.816432</td>\n",
       "      <td>0.080400</td>\n",
       "      <td>1.933158e-51</td>\n",
       "      <td>0.405944</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>ENO1</td>\n",
       "      <td>-1.080259</td>\n",
       "      <td>8.320121e-54</td>\n",
       "      <td>0.722192</td>\n",
       "      <td>0.663600</td>\n",
       "      <td>2.230981e-51</td>\n",
       "      <td>0.919522</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1763</th>\n",
       "      <td>PLAUR</td>\n",
       "      <td>-1.002734</td>\n",
       "      <td>1.306125e-53</td>\n",
       "      <td>-0.052274</td>\n",
       "      <td>0.013200</td>\n",
       "      <td>3.064496e-51</td>\n",
       "      <td>0.138416</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>ACTB</td>\n",
       "      <td>-0.746223</td>\n",
       "      <td>4.067266e-53</td>\n",
       "      <td>-0.016385</td>\n",
       "      <td>0.000125</td>\n",
       "      <td>8.482509e-51</td>\n",
       "      <td>0.003714</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1497</th>\n",
       "      <td>PFN1</td>\n",
       "      <td>-0.782788</td>\n",
       "      <td>1.129414e-51</td>\n",
       "      <td>0.465121</td>\n",
       "      <td>0.000183</td>\n",
       "      <td>2.119910e-49</td>\n",
       "      <td>0.004828</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "          gene   de_coef       de_pval   dv_coef   dv_pval        de_fdr  \\\n",
       "357   RPL15    -0.830076  4.407524e-77  0.531724  0.000182  8.272923e-74   \n",
       "1158  GAPDH    -0.648887  9.882150e-67  0.033378  0.003000  9.274397e-64   \n",
       "49    MARCKSL1 -1.277120  7.097472e-65  0.478701  0.053200  4.440651e-62   \n",
       "856   RPL10    -0.666116  9.276986e-61  0.238853  0.599800  4.353226e-58   \n",
       "1105  FAU      -0.379072  2.031012e-54  0.525444  0.010200  7.624420e-52   \n",
       "1238  RPL6     -1.343297  6.179513e-54  0.816432  0.080400  1.933158e-51   \n",
       "11    ENO1     -1.080259  8.320121e-54  0.722192  0.663600  2.230981e-51   \n",
       "1763  PLAUR    -1.002734  1.306125e-53 -0.052274  0.013200  3.064496e-51   \n",
       "716   ACTB     -0.746223  4.067266e-53 -0.016385  0.000125  8.482509e-51   \n",
       "1497  PFN1     -0.782788  1.129414e-51  0.465121  0.000183  2.119910e-49   \n",
       "\n",
       "        dv_fdr  \n",
       "357   0.004828  \n",
       "1158  0.047319  \n",
       "49    0.324209  \n",
       "856   0.889168  \n",
       "1105  0.117456  \n",
       "1238  0.405944  \n",
       "11    0.919522  \n",
       "1763  0.138416  \n",
       "716   0.003714  \n",
       "1497  0.004828  "
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_1d.sort_values('de_fdr').head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene</th>\n",
       "      <th>de_coef</th>\n",
       "      <th>de_pval</th>\n",
       "      <th>dv_coef</th>\n",
       "      <th>dv_pval</th>\n",
       "      <th>de_fdr</th>\n",
       "      <th>dv_fdr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>915</th>\n",
       "      <td>LY6E</td>\n",
       "      <td>3.497728</td>\n",
       "      <td>9.358025e-09</td>\n",
       "      <td>-3.048274</td>\n",
       "      <td>5.533871e-37</td>\n",
       "      <td>3.721401e-08</td>\n",
       "      <td>1.038708e-33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>472</th>\n",
       "      <td>CXCL10</td>\n",
       "      <td>5.699998</td>\n",
       "      <td>2.306696e-06</td>\n",
       "      <td>-3.385572</td>\n",
       "      <td>2.613791e-34</td>\n",
       "      <td>6.293122e-06</td>\n",
       "      <td>2.453042e-31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1039</th>\n",
       "      <td>IFITM3</td>\n",
       "      <td>3.458744</td>\n",
       "      <td>1.172596e-06</td>\n",
       "      <td>-2.698284</td>\n",
       "      <td>2.285528e-30</td>\n",
       "      <td>3.391313e-06</td>\n",
       "      <td>1.429979e-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>876</th>\n",
       "      <td>IDO1</td>\n",
       "      <td>3.987419</td>\n",
       "      <td>3.147780e-07</td>\n",
       "      <td>-1.980406</td>\n",
       "      <td>3.747547e-30</td>\n",
       "      <td>1.020446e-06</td>\n",
       "      <td>1.758537e-27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1527</th>\n",
       "      <td>CCL2</td>\n",
       "      <td>1.561331</td>\n",
       "      <td>1.951830e-04</td>\n",
       "      <td>-0.939252</td>\n",
       "      <td>9.723871e-27</td>\n",
       "      <td>3.385939e-04</td>\n",
       "      <td>3.650341e-24</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1421</th>\n",
       "      <td>ISG20</td>\n",
       "      <td>3.707429</td>\n",
       "      <td>3.211474e-08</td>\n",
       "      <td>-2.312789</td>\n",
       "      <td>1.797152e-24</td>\n",
       "      <td>1.196019e-07</td>\n",
       "      <td>5.622090e-22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>284</th>\n",
       "      <td>IL1RN</td>\n",
       "      <td>4.349530</td>\n",
       "      <td>4.602906e-06</td>\n",
       "      <td>-1.750747</td>\n",
       "      <td>4.047650e-18</td>\n",
       "      <td>1.172273e-05</td>\n",
       "      <td>1.085348e-15</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1300</th>\n",
       "      <td>PSME2</td>\n",
       "      <td>0.883725</td>\n",
       "      <td>2.048874e-07</td>\n",
       "      <td>-0.813046</td>\n",
       "      <td>6.302435e-17</td>\n",
       "      <td>6.770661e-07</td>\n",
       "      <td>1.478709e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>37</th>\n",
       "      <td>IFI6</td>\n",
       "      <td>2.811940</td>\n",
       "      <td>7.646498e-08</td>\n",
       "      <td>-1.910721</td>\n",
       "      <td>1.204907e-16</td>\n",
       "      <td>2.692773e-07</td>\n",
       "      <td>2.512900e-14</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>217</th>\n",
       "      <td>RSAD2</td>\n",
       "      <td>4.909576</td>\n",
       "      <td>5.487289e-07</td>\n",
       "      <td>-2.470896</td>\n",
       "      <td>6.276557e-16</td>\n",
       "      <td>1.702420e-06</td>\n",
       "      <td>1.178110e-13</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        gene   de_coef       de_pval   dv_coef       dv_pval        de_fdr  \\\n",
       "915   LY6E    3.497728  9.358025e-09 -3.048274  5.533871e-37  3.721401e-08   \n",
       "472   CXCL10  5.699998  2.306696e-06 -3.385572  2.613791e-34  6.293122e-06   \n",
       "1039  IFITM3  3.458744  1.172596e-06 -2.698284  2.285528e-30  3.391313e-06   \n",
       "876   IDO1    3.987419  3.147780e-07 -1.980406  3.747547e-30  1.020446e-06   \n",
       "1527  CCL2    1.561331  1.951830e-04 -0.939252  9.723871e-27  3.385939e-04   \n",
       "1421  ISG20   3.707429  3.211474e-08 -2.312789  1.797152e-24  1.196019e-07   \n",
       "284   IL1RN   4.349530  4.602906e-06 -1.750747  4.047650e-18  1.172273e-05   \n",
       "1300  PSME2   0.883725  2.048874e-07 -0.813046  6.302435e-17  6.770661e-07   \n",
       "37    IFI6    2.811940  7.646498e-08 -1.910721  1.204907e-16  2.692773e-07   \n",
       "217   RSAD2   4.909576  5.487289e-07 -2.470896  6.276557e-16  1.702420e-06   \n",
       "\n",
       "            dv_fdr  \n",
       "915   1.038708e-33  \n",
       "472   2.453042e-31  \n",
       "1039  1.429979e-27  \n",
       "876   1.758537e-27  \n",
       "1527  3.650341e-24  \n",
       "1421  5.622090e-22  \n",
       "284   1.085348e-15  \n",
       "1300  1.478709e-14  \n",
       "37    2.512900e-14  \n",
       "217   1.178110e-13  "
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_1d.sort_values('dv_fdr').head(10)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Perform 2D hypothesis testing\n",
    "\n",
    "For differential co-variability testing, we can specify which genes you want to perform HT on. It takes a list of pairs of genes, where each element in the list is a tuple. Here, we focus on 1 transcription factor and their correlations to rest of the transcriptome. \n",
    "\n",
    "Similar to the 1D case, 2D hypothesis testing scales with the number of pairs of genes to test. If you have a smaller set of candidate genes, it will run faster."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "import itertools"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "gene_pairs = list(itertools.product(['IRF7'], adata.var.index.tolist()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "memento.compute_2d_moments(adata, gene_pairs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "[Parallel(n_jobs=6)]: Using backend LokyBackend with 6 concurrent workers.\n",
      "[Parallel(n_jobs=6)]: Done  20 tasks      | elapsed:    2.8s\n",
      "[Parallel(n_jobs=6)]: Done 116 tasks      | elapsed:   14.8s\n",
      "[Parallel(n_jobs=6)]: Done 276 tasks      | elapsed:   35.9s\n",
      "[Parallel(n_jobs=6)]: Done 500 tasks      | elapsed:  1.1min\n",
      "[Parallel(n_jobs=6)]: Done 788 tasks      | elapsed:  1.7min\n",
      "[Parallel(n_jobs=6)]: Done 1140 tasks      | elapsed:  2.5min\n",
      "[Parallel(n_jobs=6)]: Done 1556 tasks      | elapsed:  3.4min\n",
      "[Parallel(n_jobs=6)]: Done 1876 out of 1876 | elapsed:  4.1min finished\n"
     ]
    }
   ],
   "source": [
    "memento.ht_2d_moments(\n",
    "    adata, \n",
    "    formula_like='1 + stim', \n",
    "    cov_column='stim', \n",
    "    num_cpus=6, \n",
    "    num_boot=5000)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "result_2d = memento.get_2d_ht_result(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gene_1</th>\n",
       "      <th>gene_2</th>\n",
       "      <th>corr_coef</th>\n",
       "      <th>corr_pval</th>\n",
       "      <th>corr_fdr</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>574</th>\n",
       "      <td>IRF7</td>\n",
       "      <td>CD74</td>\n",
       "      <td>0.316293</td>\n",
       "      <td>0.000123</td>\n",
       "      <td>0.073478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1815</th>\n",
       "      <td>IRF7</td>\n",
       "      <td>SDF2L1</td>\n",
       "      <td>0.304159</td>\n",
       "      <td>0.000181</td>\n",
       "      <td>0.073478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>104</th>\n",
       "      <td>IRF7</td>\n",
       "      <td>GCLM</td>\n",
       "      <td>0.396597</td>\n",
       "      <td>0.000283</td>\n",
       "      <td>0.073478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>716</th>\n",
       "      <td>IRF7</td>\n",
       "      <td>ACTB</td>\n",
       "      <td>0.272642</td>\n",
       "      <td>0.000334</td>\n",
       "      <td>0.073478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>638</th>\n",
       "      <td>IRF7</td>\n",
       "      <td>HLA-DRA</td>\n",
       "      <td>0.252754</td>\n",
       "      <td>0.000473</td>\n",
       "      <td>0.073478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>158</th>\n",
       "      <td>IRF7</td>\n",
       "      <td>LMNA</td>\n",
       "      <td>0.306795</td>\n",
       "      <td>0.000560</td>\n",
       "      <td>0.073478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1108</th>\n",
       "      <td>IRF7</td>\n",
       "      <td>MALAT1</td>\n",
       "      <td>0.249095</td>\n",
       "      <td>0.000571</td>\n",
       "      <td>0.073478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>493</th>\n",
       "      <td>IRF7</td>\n",
       "      <td>ANXA5</td>\n",
       "      <td>0.275887</td>\n",
       "      <td>0.000642</td>\n",
       "      <td>0.073478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>211</th>\n",
       "      <td>IRF7</td>\n",
       "      <td>GPR137B</td>\n",
       "      <td>0.397522</td>\n",
       "      <td>0.000659</td>\n",
       "      <td>0.073478</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>626</th>\n",
       "      <td>IRF7</td>\n",
       "      <td>HLA-C</td>\n",
       "      <td>0.268576</td>\n",
       "      <td>0.000686</td>\n",
       "      <td>0.073478</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "     gene_1   gene_2  corr_coef  corr_pval  corr_fdr\n",
       "574   IRF7   CD74     0.316293   0.000123   0.073478\n",
       "1815  IRF7   SDF2L1   0.304159   0.000181   0.073478\n",
       "104   IRF7   GCLM     0.396597   0.000283   0.073478\n",
       "716   IRF7   ACTB     0.272642   0.000334   0.073478\n",
       "638   IRF7   HLA-DRA  0.252754   0.000473   0.073478\n",
       "158   IRF7   LMNA     0.306795   0.000560   0.073478\n",
       "1108  IRF7   MALAT1   0.249095   0.000571   0.073478\n",
       "493   IRF7   ANXA5    0.275887   0.000642   0.073478\n",
       "211   IRF7   GPR137B  0.397522   0.000659   0.073478\n",
       "626   IRF7   HLA-C    0.268576   0.000686   0.073478"
      ]
     },
     "execution_count": 33,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_2d.sort_values('corr_pval').head(10)"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
