{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c95c73d7",
   "metadata": {},
   "source": [
    "### ATAC enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "0d4f76cb",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "######################\n",
      "#### bedr v1.0.7 ####\n",
      "######################\n",
      "\n",
      "checking binary availability...\n",
      "  * Checking path for bedtools... PASS\n",
      "    /home/ssm-user/anaconda3/envs/single_cell/bin/bedtools\n",
      "  * Checking path for bedops... FAIL\n",
      "  * Checking path for tabix... FAIL\n",
      "tests and examples will be skipped on R CMD check if binaries are missing\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(data.table)\n",
    "# library(qvalue)\n",
    "library(bedr)\n",
    "options(scipen = 999)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "da896bb3",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "for (ct in c('B', 'T8', 'T4', 'NK', 'ncM', 'cM')){\n",
    "\n",
    "    result_path = paste(\"/data_volume/memento/lupus/full_analysis/memento/hits/eur_\", ct,\".csv\", sep=\"\")\n",
    "#     result_path = paste(\"/data_volume/memento/lupus/full_analysis/mateqtl/outputs/asian_\", ct,\".csv\", sep=\"\")\n",
    "\n",
    "    #res_raw=data.frame(fread(args[1]))\n",
    "    res_raw=data.frame(fread(result_path))\n",
    "    #emp=fread(args[2])\n",
    "    #delta=fread(args[3])\n",
    "    out= paste(\"/data_volume/memento/lupus/atac_enrichment/output/eur_\", ct, \"_atac_mem.txt\", sep=\"\")\n",
    "\n",
    "    \n",
    "    expr_genes_path =  paste(\"/data_volume/memento/lupus/atac_enrichment/expr_genes/eur_\", ct, \".csv\", sep=\"\")\n",
    "    expr_genes=read.table(expr_genes_path,header=F)\n",
    "\n",
    "    print('gene')\n",
    "    print(head(expr_genes))\n",
    "    print(dim(expr_genes))\n",
    "\n",
    "\n",
    "    #print(head(emp))\n",
    "    #emp$FDR=qvalue(emp$Empirical)$qvalues\n",
    "    #thresh=max(emp$Pvalue[which(emp$FDR < 0.1)])\n",
    "\n",
    "    #run if not corrected by celltype\n",
    "    #fdr=qvalue(res_raw$pvalue)\n",
    "    #print(fdr)\n",
    "    #res_raw$FDR=fdr$qvalues\n",
    "\n",
    "    #filter FDR\n",
    "    #res=res_raw[res_raw$FDR <= 0.1,]\n",
    "    #snps gene statistic pvalue FDR beta qvalue\n",
    "\n",
    "\n",
    "\n",
    "    colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\")\n",
    "    #colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\",\"qvalue\")\n",
    "    #res=res_raw\n",
    "    #filter for expressed genes\n",
    "#     print(head(res_raw))\n",
    "#     print(dim(res_raw))\n",
    "#     res_expr=res_raw[res_raw$gene %in% expr_genes$V1,]\n",
    "#     print(dim((res_expr)))\n",
    "#     res_noexpr=res_raw[!(res_raw$gene %in% expr_genes$V1),]\n",
    "\n",
    "#     print(dim(res_expr))\n",
    "#     print(head(res_expr))\n",
    "#     print(head(res_noexpr))\n",
    "\n",
    "#     num_no=length(res_noexpr$gene)\n",
    "#     nullp=runif(num_no, min = 0, max = 1)\n",
    "#     res_noexpr$p.value=nullp\n",
    "#     res=rbind(res_expr,res_noexpr)\n",
    "#     print(head(res_expr))\n",
    "#     print(dim(res_expr))\n",
    "#     print(dim(res))\n",
    "#     print(tail(res))\n",
    "    #atac=fread('/ye/yelabstore2/10x.lupus/process.scRNAseq/eqtls/S9_lineage_groups.txt.gz')\n",
    "    atac=fread('/data_volume/memento/lupus/atac_enrichment/simple_sorted_atac_lineage_groups.txt.gz')\n",
    "    #names(atac)=c('chr', 'start', 'end', 'group')\n",
    "    #atac.bed=data.frame(chr=atac$chr, start=atac$start, end=atac$stop, group=atac$group)\n",
    "    atac.groups=names(table(atac$group))\n",
    "    #atac.sorted=bedr.sort.region(atac$chr,atac$start,atac$stop)\n",
    "    res =res_raw\n",
    "\n",
    "    res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "    res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "\n",
    "    #res$chr=paste('chr', sapply(strsplit(res$snp, ':'), '[', 1), sep='')\n",
    "    #res$pos=as.numeric(sapply(strsplit(res$snp, ':'), '[', 2))\n",
    "    res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "    res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "    #res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "    #res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "    #res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "    #res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "    all.df=NULL\n",
    "    for(g in atac.groups){\n",
    "        print(g)\n",
    "        peaks=atac[which(atac$group == g), ]\n",
    "        peaks=paste(peaks$chr, ':', peaks$start, '-', peaks$stop, sep='')\n",
    "        overlap.sig.snps=bedr(input=list(a=res.bed.sorted, b=peaks), method='intersect')\n",
    "\n",
    "        print(\"here~~~~~~~~~\")\n",
    "\n",
    "        in_peak=as.numeric(res.bed.sorted %in% overlap.sig.snps)\n",
    "\n",
    "        pvals=res$\"p.value\"[match(res.bed.sorted, res$id)]\n",
    "        #pvals=res$\"pvalue\"[match(res.bed.sorted, res$id)]\n",
    "        #pvals=res$\"p-value\"[match(res.bed.sorted, res$id)]\n",
    "        mw_res=wilcox.test(pvals ~ in_peak, alternative='greater')\n",
    "        df=data.frame(W=mw_res$statistic, pval=mw_res$p.value, group=g)\n",
    "        all.df=rbind(all.df, df)\n",
    "        print(all.df)\n",
    "    }\n",
    "\n",
    "    write.table(all.df, file=out, row.names=F, col.names=T, quote=F, sep='\\t')\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "28366c85",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "0"
      ],
      "text/latex": [
       "0"
      ],
      "text/markdown": [
       "0"
      ],
      "text/plain": [
       "[1] 0"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "sum(in_peak)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e55daad5",
   "metadata": {},
   "source": [
    "### Gracie attempt 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "42726655",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Warning message in fread(\"/data_volume/memento/lupus/gracie_eqtls/nk_metasoft_out.txt\"):\n",
      "“Detected 18 column names but the data has 23 columns (i.e. invalid file). Added 5 extra default column names at the end.”\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"gene\"\n",
      "                  RSID STUDY  p.value     BETA_FE    STD_FE PVALUE_RE\n",
      "1 10:100043981_PYROXD2     2 0.499939  0.05168100 0.0766115  0.569163\n",
      "2 10:100044928_PYROXD2     3 0.481351 -0.03292740 0.0467632  0.533360\n",
      "3 10:100045000_PYROXD2     3 0.333619 -0.03689180 0.0381567  0.631411\n",
      "4 10:100045147_PYROXD2     3 0.498166 -0.02550860 0.0376579  0.748616\n",
      "5 10:100045166_PYROXD2     2 0.873178  0.01125270 0.0704954  0.955431\n",
      "6 10:100045813_PYROXD2     2 0.909779 -0.00787023 0.0694531  0.965707\n",
      "      BETA_RE    STD_RE PVALUE_RE2 STAT1_RE2 STAT2_RE2 PVALUE_BE I_SQUARE\n",
      "1  0.04935050 0.0866885  0.5359850 0.4550660  0.000000        NA  21.4242\n",
      "2 -0.03152080 0.0506046  0.5344880 0.4958000  0.000000        NA  14.0455\n",
      "3 -0.02936590 0.0612119  0.2534790 0.9347980  0.597342        NA  60.6759\n",
      "4 -0.01923900 0.0600347  0.3559570 0.4588410  0.573694        NA  60.2236\n",
      "5 -0.01125750  0.201428  0.0812551 0.0254796  3.236490        NA  87.5518\n",
      "6  0.00850442  0.197812  0.0812569 0.0128408  3.249090        NA  87.5629\n",
      "        Q   PVALUE_Q TAU_SQUARE PVALUES_OF_STUDIES(Tab_delimitered)\n",
      "1 1.27266 0.25926900 0.00325131                                  NA\n",
      "2 2.32681 0.31242000 0.00109033                           0.8309240\n",
      "3 5.08594 0.07863270 0.00681595                           0.0208621\n",
      "4 5.02811 0.08093940 0.00650758                           0.0367613\n",
      "5 8.03331 0.00459249  0.0710629                           0.0463973\n",
      "6 8.04047 0.00457436  0.0685353                           0.0463973\n",
      "  MVALUES_OF_STUDIES(Tab_delimitered)        NA NA NA NA NA          SNP\n",
      "1                            0.210517 0.6892910 NA NA NA NA 10:100043981\n",
      "2                            0.707767 0.1044330 NA NA NA NA 10:100044928\n",
      "3                            0.411946 0.9232660 NA NA NA NA 10:100045000\n",
      "4                            0.290730 0.9232660 NA NA NA NA 10:100045147\n",
      "5                                  NA 0.0430949 NA NA NA NA 10:100045166\n",
      "6                                  NA 0.0432346 NA NA NA NA 10:100045813\n",
      "     gene\n",
      "1 PYROXD2\n",
      "2 PYROXD2\n",
      "3 PYROXD2\n",
      "4 PYROXD2\n",
      "5 PYROXD2\n",
      "6 PYROXD2\n",
      "[1] 3379632      25\n",
      "[1] \"expr\"\n",
      "[1] \"no gene\"\n",
      "[1] 3379632      25\n",
      "                     RSID STUDY  p.value     BETA_FE    STD_FE PVALUE_RE\n",
      "3379627 9:99802653_HIATL2     2 0.609609  0.02577020 0.0504672  0.609609\n",
      "3379628 9:99802796_HIATL2     2 0.609609  0.02577020 0.0504672  0.609609\n",
      "3379629 9:99804997_HIATL2     2 0.918560 -0.00555366 0.0543158  0.851717\n",
      "3379630 9:99805079_HIATL2     2 0.609609  0.02577020 0.0504672  0.609609\n",
      "3379631 9:99806169_HIATL2     2 0.609609  0.02577020 0.0504672  0.609609\n",
      "3379632 9:99806787_HIATL2     2 0.609609  0.02577020 0.0504672  0.609609\n",
      "           BETA_RE    STD_RE PVALUE_RE2 STAT1_RE2 STAT2_RE2 PVALUE_BE I_SQUARE\n",
      "3379627  0.0257702 0.0504672   0.644582 0.2607450   0.00000        NA   0.0000\n",
      "3379628  0.0257702 0.0504672   0.644582 0.2607450   0.00000        NA   0.0000\n",
      "3379629 -0.0262363  0.140355   0.165065 0.0104546   2.09906        NA  84.5664\n",
      "3379630  0.0257702 0.0504672   0.644582 0.2607450   0.00000        NA   0.0000\n",
      "3379631  0.0257702 0.0504672   0.644582 0.2607450   0.00000        NA   0.0000\n",
      "3379632  0.0257702 0.0504672   0.644582 0.2607450   0.00000        NA   0.0000\n",
      "               Q  PVALUE_Q TAU_SQUARE PVALUES_OF_STUDIES(Tab_delimitered)\n",
      "3379627 0.224914 0.6353210    0.00000                            0.486124\n",
      "3379628 0.224914 0.6353210    0.00000                            0.486124\n",
      "3379629 6.479390 0.0109133  0.0333423                            0.119397\n",
      "3379630 0.224914 0.6353210    0.00000                            0.486124\n",
      "3379631 0.224914 0.6353210    0.00000                            0.486124\n",
      "3379632 0.224914 0.6353210    0.00000                            0.486124\n",
      "        MVALUES_OF_STUDIES(Tab_delimitered) NA NA NA NA NA        SNP   gene\n",
      "3379627                           0.9810430 NA NA NA NA NA 9:99802653 HIATL2\n",
      "3379628                           0.9810430 NA NA NA NA NA 9:99802796 HIATL2\n",
      "3379629                           0.0437905 NA NA NA NA NA 9:99804997 HIATL2\n",
      "3379630                           0.9810430 NA NA NA NA NA 9:99805079 HIATL2\n",
      "3379631                           0.9810430 NA NA NA NA NA 9:99806169 HIATL2\n",
      "3379632                           0.9810430 NA NA NA NA NA 9:99806787 HIATL2\n",
      "SORTING\n",
      "VALIDATE REGIONS\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      "[1] \"B\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/Rtmp6UEF5N/a_33e96058f3a9.bed -b /tmp/Rtmp6UEF5N/b_33e91e21fcda.bed \n",
      "[1] \"EPI\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/Rtmp6UEF5N/a_33e912292ac9.bed -b /tmp/Rtmp6UEF5N/b_33e9d259d3b.bed \n",
      "[1] \"myeloid\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/Rtmp6UEF5N/a_33e970bd3a06.bed -b /tmp/Rtmp6UEF5N/b_33e96bf615b.bed \n",
      "[1] \"nk\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/Rtmp6UEF5N/a_33e93106325c.bed -b /tmp/Rtmp6UEF5N/b_33e95a883a49.bed \n",
      "[1] \"open\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/Rtmp6UEF5N/a_33e97a685529.bed -b /tmp/Rtmp6UEF5N/b_33e93f5ffec6.bed \n",
      "[1] \"progenitor\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/Rtmp6UEF5N/a_33e97a0ddd5d.bed -b /tmp/Rtmp6UEF5N/b_33e911281bd1.bed \n",
      "[1] \"T\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/Rtmp6UEF5N/a_33e91e864f52.bed -b /tmp/Rtmp6UEF5N/b_33e92f5d9320.bed \n",
      "[1] \"thymocyte\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/Rtmp6UEF5N/a_33e9172c0e6e.bed -b /tmp/Rtmp6UEF5N/b_33e966c1be64.bed \n"
     ]
    }
   ],
   "source": [
    "library(data.table)\n",
    "library(qvalue)\n",
    "library(bedr)\n",
    "options(scipen = 999)\n",
    "# args=commandArgs(TRUE)\n",
    "\n",
    "res_raw=data.frame(fread('/data_volume/memento/lupus/gracie_eqtls/nk_metasoft_out.txt'))\n",
    "#res_raw=data.frame(fread(args[1]))[-1]\n",
    "#emp=fread(args[2])\n",
    "#delta=fread(args[3])\n",
    "out='gracie_nk.out'\n",
    "#expr_genes=read.table(args[3],header=F)\n",
    "\n",
    "print('gene')\n",
    "#print(head(expr_genes))\n",
    "#print(dim(expr_genes))\n",
    "#print(head(emp))\n",
    "#emp$FDR=qvalue(emp$Empirical)$qvalues\n",
    "#thresh=max(emp$Pvalue[which(emp$FDR < 0.1)])\n",
    "\n",
    "#run if not corrected by celltype\n",
    "#fdr=qvalue(res_raw$pvalue)\n",
    "#print(fdr)\n",
    "#res_raw$FDR=fdr$qvalues\n",
    "\n",
    "#filter FDR\n",
    "#res=res_raw[res_raw$FDR <= 0.1,]\n",
    "#snps gene statistic pvalue FDR beta qvalue\n",
    "\n",
    "\n",
    "#colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\")\n",
    "#take in metasoft data, create atac enrichments from fixed effects\n",
    "colnames(res_raw)=c(\"RSID\",\"STUDY\",\"p.value\",\"BETA_FE\",\"STD_FE\",\"PVALUE_RE\",\"BETA_RE\",\"STD_RE\",\"PVALUE_RE2\",\"STAT1_RE2\",\"STAT2_RE2\",\"PVALUE_BE\",\"I_SQUARE\",\"Q\",\"PVALUE_Q\",\"TAU_SQUARE\",\"PVALUES_OF_STUDIES(Tab_delimitered)\",\"MVALUES_OF_STUDIES(Tab_delimitered)\")\n",
    "res_raw$SNP=sapply(strsplit(res_raw$RSID, '_'), '[', 1)\n",
    "res_raw$gene=sapply(strsplit(res_raw$RSID, '_'), '[', 2)\n",
    "\n",
    "#colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\",\"qvalue\")\n",
    "#res=res_raw\n",
    "#filter for expressed genes\n",
    "print(head(res_raw))\n",
    "print(dim(res_raw))\n",
    "#res_expr=res_raw[res_raw$gene %in% expr_genes$V1,]\n",
    "#print(dim((res_expr)))\n",
    "#res_noexpr=res_raw[!(res_raw$gene %in% expr_genes$V1),]\n",
    "\n",
    "print('expr')\n",
    "#print(dim(res_expr))\n",
    "#print(head(res_expr))\n",
    "print('no gene')\n",
    "#print(head(res_noexpr))\n",
    "\n",
    "#num_no=length(res_noexpr$gene)\n",
    "#print(num_no)\n",
    "#nullp=runif(num_no, min = 0, max = 1)\n",
    "#nullp=rep(1,num_no)\n",
    "#res_noexpr$p.value=nullp\n",
    "#res=rbind(res_expr,res_noexpr)\n",
    "#print(head(res_expr))\n",
    "#print(dim(res_expr))\n",
    "res=res_raw\n",
    "print(dim(res))\n",
    "print(tail(res))\n",
    "#atac=fread('/ye/yelabstore2/10x.lupus/process.scRNAseq/eqtls/S9_lineage_groups.txt.gz')\n",
    "#atac=fread('simple_sorted_atac_lineage_groups2.txt.gz')\n",
    "atac=fread('/data_volume/memento/lupus/atac_enrichment/sorted_simple_atac_lineage_groups3.bed.gz')\n",
    "\n",
    "#names(atac)=c('chr', 'start', 'end', 'group')\n",
    "#atac.bed=data.frame(chr=atac$chr, start=atac$start, end=atac$stop, group=atac$group)\n",
    "atac.groups=names(table(atac$group))\n",
    "#atac.sorted=bedr.sort.region(atac$chr,atac$start,atac$stop)\n",
    "\n",
    "res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "\n",
    "#res$chr=paste('chr', sapply(strsplit(res$snp, ':'), '[', 1), sep='')\n",
    "#res$pos=as.numeric(sapply(strsplit(res$snp, ':'), '[', 2))\n",
    "res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "#res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "#res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "#res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "#res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "all.df=NULL\n",
    "for(g in atac.groups){\n",
    "    print(g)\n",
    "    peaks=atac[which(atac$group == g), ]\n",
    "    peaks=paste(peaks$chr, ':', peaks$start, '-', peaks$stop, sep='')\n",
    "    overlap.sig.snps=bedr(input=list(a=res.bed.sorted, b=peaks), method='intersect')\n",
    "    in_peak=as.numeric(res.bed.sorted %in% overlap.sig.snps)\n",
    "    pvals=res$\"p.value\"[match(res.bed.sorted, res$id)]\n",
    "    #pvals=res$\"pvalue\"[match(res.bed.sorted, res$id)]\n",
    "    #pvals=res$\"p-value\"[match(res.bed.sorted, res$id)]\n",
    "    mw_res=wilcox.test(pvals ~ in_peak, alternative='greater')\n",
    "    df=data.frame(W=mw_res$statistic, pval=mw_res$p.value, group=g)\n",
    "    all.df=rbind(all.df, df)   \n",
    "}\n",
    "\n",
    "write.table(all.df, file=out, row.names=F, col.names=T, quote=F, sep='\\t')\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "ec6cfd8f",
   "metadata": {},
   "source": [
    "### Copy gracie's attempt with memento  eqtls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "9d6bcc71",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (ct in c('T4', 'cM', 'ncM', 'T8', 'B', 'NK')){\n",
    "print(ct)\n",
    "library(data.table)\n",
    "library(qvalue)\n",
    "library(bedr)\n",
    "options(scipen = 999)\n",
    "# args=commandArgs(TRUE)\n",
    "\n",
    "res_raw=data.frame(fread(paste('/data_volume/memento/lupus/full_analysis/memento/gracie/eur_',ct, '.csv', sep='')))\n",
    "#res_raw=data.frame(fread(args[1]))[-1]\n",
    "#emp=fread(args[2])\n",
    "#delta=fread(args[3])\n",
    "out=paste('/data_volume/memento/lupus/atac_enrichment/gracie/', 'eur_matqetl_',ct, '.out', sep='')\n",
    "#expr_genes=read.table(args[3],header=F)\n",
    "\n",
    "print('gene')\n",
    "#print(head(expr_genes))\n",
    "#print(dim(expr_genes))\n",
    "#print(head(emp))\n",
    "#emp$FDR=qvalue(emp$Empirical)$qvalues\n",
    "#thresh=max(emp$Pvalue[which(emp$FDR < 0.1)])\n",
    "\n",
    "#run if not corrected by celltype\n",
    "#fdr=qvalue(res_raw$pvalue)\n",
    "#print(fdr)\n",
    "#res_raw$FDR=fdr$qvalues\n",
    "\n",
    "#filter FDR\n",
    "#res=res_raw[res_raw$FDR <= 0.1,]\n",
    "#snps gene statistic pvalue FDR beta qvalue\n",
    "\n",
    "\n",
    "#colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\")\n",
    "#take in metasoft data, create atac enrichments from fixed effects\n",
    "colnames(res_raw)=c('SNP', 'gene', 'statistic', 'p.value', 'FDR', 'beta')\n",
    "# res_raw$gene=sapply(strsplit(res_raw$RSID, '_'), '[', 2)\n",
    "\n",
    "#colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\",\"qvalue\")\n",
    "#res=res_raw\n",
    "#filter for expressed genes\n",
    "print(head(res_raw))\n",
    "print(dim(res_raw))\n",
    "#res_expr=res_raw[res_raw$gene %in% expr_genes$V1,]\n",
    "#print(dim((res_expr)))\n",
    "#res_noexpr=res_raw[!(res_raw$gene %in% expr_genes$V1),]\n",
    "\n",
    "print('expr')\n",
    "#print(dim(res_expr))\n",
    "#print(head(res_expr))\n",
    "print('no gene')\n",
    "#print(head(res_noexpr))\n",
    "\n",
    "#num_no=length(res_noexpr$gene)\n",
    "#print(num_no)\n",
    "#nullp=runif(num_no, min = 0, max = 1)\n",
    "#nullp=rep(1,num_no)\n",
    "#res_noexpr$p.value=nullp\n",
    "#res=rbind(res_expr,res_noexpr)\n",
    "#print(head(res_expr))\n",
    "#print(dim(res_expr))\n",
    "res=res_raw\n",
    "print(dim(res))\n",
    "print(tail(res))\n",
    "#atac=fread('/ye/yelabstore2/10x.lupus/process.scRNAseq/eqtls/S9_lineage_groups.txt.gz')\n",
    "#atac=fread('simple_sorted_atac_lineage_groups2.txt.gz')\n",
    "atac=fread('/data_volume/memento/lupus/atac_enrichment/sorted_simple_atac_lineage_groups3.bed.gz')\n",
    "\n",
    "#names(atac)=c('chr', 'start', 'end', 'group')\n",
    "#atac.bed=data.frame(chr=atac$chr, start=atac$start, end=atac$stop, group=atac$group)\n",
    "atac.groups=names(table(atac$group))\n",
    "#atac.sorted=bedr.sort.region(atac$chr,atac$start,atac$stop)\n",
    "\n",
    "res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "\n",
    "#res$chr=paste('chr', sapply(strsplit(res$snp, ':'), '[', 1), sep='')\n",
    "#res$pos=as.numeric(sapply(strsplit(res$snp, ':'), '[', 2))\n",
    "res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "#res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "#res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "#res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "#res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "all.df=NULL\n",
    "for(g in atac.groups){\n",
    "    print(g)\n",
    "    peaks=atac[which(atac$group == g), ]\n",
    "    peaks=paste(peaks$chr, ':', peaks$start, '-', peaks$stop, sep='')\n",
    "    overlap.sig.snps=bedr(input=list(a=res.bed.sorted, b=peaks), method='intersect')\n",
    "    in_peak=as.numeric(res.bed.sorted %in% overlap.sig.snps)\n",
    "    pvals=res$\"p.value\"[match(res.bed.sorted, res$id)]\n",
    "    #pvals=res$\"pvalue\"[match(res.bed.sorted, res$id)]\n",
    "    #pvals=res$\"p-value\"[match(res.bed.sorted, res$id)]\n",
    "    mw_res=wilcox.test(pvals ~ in_peak, alternative='greater')\n",
    "    df=data.frame(W=mw_res$statistic, pval=mw_res$p.value, group=g)\n",
    "    all.df=rbind(all.df, df)   \n",
    "}\n",
    "\n",
    "write.table(all.df, file=out, row.names=F, col.names=T, quote=F, sep='\\t')\n",
    "\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "dc87e12c",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "d41142fc",
   "metadata": {},
   "source": [
    "### Copy gracie's attempt with mateqtl  eqtls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "35c43612",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"T4\"\n",
      "[1] \"gene\"\n",
      "          SNP  gene      beta statistic\n",
      "1 22:42475703 SMDT1 -0.213625  -29.4415\n",
      "2 22:42476754 SMDT1 -0.213625  -29.4415\n",
      "3 22:42484466 SMDT1 -0.213625  -29.4415\n",
      "4 22:42487900 SMDT1 -0.213625  -29.4415\n",
      "5 22:42488160 SMDT1 -0.213625  -29.4415\n",
      "6 22:42490600 SMDT1 -0.213625  -29.4415\n",
      "                                      p.value\n",
      "1 0.00000000000000000000000000000000006144902\n",
      "2 0.00000000000000000000000000000000006144902\n",
      "3 0.00000000000000000000000000000000006144902\n",
      "4 0.00000000000000000000000000000000006144902\n",
      "5 0.00000000000000000000000000000000006144902\n",
      "6 0.00000000000000000000000000000000006144902\n",
      "                                     FDR\n",
      "1 0.000000000000000000000000000002600136\n",
      "2 0.000000000000000000000000000002600136\n",
      "3 0.000000000000000000000000000002600136\n",
      "4 0.000000000000000000000000000002600136\n",
      "5 0.000000000000000000000000000002600136\n",
      "6 0.000000000000000000000000000002600136\n",
      "[1] 1015529       6\n",
      "[1] \"expr\"\n",
      "[1] \"no gene\"\n",
      "[1] 1015529       6\n",
      "                SNP    gene               beta        statistic   p.value\n",
      "1015524 17:36827523   CISD3  0.000000014149724  0.0000064924405 0.9999948\n",
      "1015525 15:89190361     AEN -0.000000002385297 -0.0000023227813 0.9999982\n",
      "1015526 15:89190362     AEN -0.000000002385297 -0.0000023227813 0.9999982\n",
      "1015527 4:141246800    SCOC  0.000000004408019  0.0000019498547 0.9999985\n",
      "1015528  19:5848000 NDUFA11 -0.000000011610813 -0.0000015254800 0.9999988\n",
      "1015529  5:37421599   WDR70 -0.000000001324545 -0.0000006029566 0.9999995\n",
      "              FDR\n",
      "1015524 0.9999995\n",
      "1015525 0.9999995\n",
      "1015526 0.9999995\n",
      "1015527 0.9999995\n",
      "1015528 0.9999995\n",
      "1015529 0.9999995\n",
      "SORTING\n",
      "VALIDATE REGIONS\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      "[1] \"B\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b42714f4e.bed -b /tmp/RtmpRxRHZx/b_1b3bac908ac.bed \n",
      "[1] \"EPI\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b49483688.bed -b /tmp/RtmpRxRHZx/b_1b3ba8e5422.bed \n",
      "[1] \"myeloid\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b4f824b42.bed -b /tmp/RtmpRxRHZx/b_1b3b1ca6d9f3.bed \n",
      "[1] \"nk\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b421fdb55.bed -b /tmp/RtmpRxRHZx/b_1b3b277e02ba.bed \n",
      "[1] \"open\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b7fa1f132.bed -b /tmp/RtmpRxRHZx/b_1b3b640297da.bed \n",
      "[1] \"progenitor\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b4afce6e7.bed -b /tmp/RtmpRxRHZx/b_1b3b3702163f.bed \n",
      "[1] \"T\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b66962738.bed -b /tmp/RtmpRxRHZx/b_1b3b1248acbb.bed \n",
      "[1] \"thymocyte\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b28b44148.bed -b /tmp/RtmpRxRHZx/b_1b3b65b771db.bed \n",
      "[1] \"cM\"\n",
      "[1] \"gene\"\n",
      "          SNP   gene       beta statistic\n",
      "1 12:56435929  RPS26 -0.5251927 -26.96663\n",
      "2 12:56401085  RPS26 -0.5181452 -24.49774\n",
      "3 12:56470625  RPS26 -0.4879132 -16.80164\n",
      "4 22:37961266 LGALS2 -0.4233050 -16.14519\n",
      "5 22:37961353 LGALS2 -0.4233050 -16.14519\n",
      "6 12:56384804  RPS26  0.4759743  16.04180\n",
      "                                    p.value                                 FDR\n",
      "1 0.000000000000000000000000000000005233315 0.000000000000000000000000006451814\n",
      "2 0.000000000000000000000000000000635741155 0.000000000000000000000000391881973\n",
      "3 0.000000000000000000000042019578170336703 0.000000000000000017267735551209001\n",
      "4 0.000000000000000000000254484968696485001 0.000000000000000052313281475326402\n",
      "5 0.000000000000000000000254484968696485001 0.000000000000000052313281475326402\n",
      "6 0.000000000000000000000339466556191712994 0.000000000000000052313281475326402\n",
      "[1] 1232835       6\n",
      "[1] \"expr\"\n",
      "[1] \"no gene\"\n",
      "[1] 1232835       6\n",
      "                 SNP     gene               beta       statistic   p.value\n",
      "1232830   19:2198845    DOT1L  0.000000009905226  0.000009289032 0.9999926\n",
      "1232831 10:101328970 SLC25A28  0.000000010658022  0.000008548879 0.9999932\n",
      "1232832  19:50307344    MED25 -0.000000007642250 -0.000005385730 0.9999957\n",
      "1232833   6:16757006    ATXN1  0.000000004402850  0.000002004162 0.9999984\n",
      "1232834   6:16758345    ATXN1  0.000000004402850  0.000002004162 0.9999984\n",
      "1232835  18:11932798    MPPE1  0.000000006693548  0.000001957394 0.9999984\n",
      "              FDR\n",
      "1232830 0.9999965\n",
      "1232831 0.9999965\n",
      "1232832 0.9999982\n",
      "1232833 0.9999984\n",
      "1232834 0.9999984\n",
      "1232835 0.9999984\n",
      "SORTING\n",
      "VALIDATE REGIONS\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      "[1] \"B\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b59639fd8.bed -b /tmp/RtmpRxRHZx/b_1b3b2efa1dfe.bed \n",
      "[1] \"EPI\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b1371e87d.bed -b /tmp/RtmpRxRHZx/b_1b3b3a5b024a.bed \n",
      "[1] \"myeloid\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b6c0d24c4.bed -b /tmp/RtmpRxRHZx/b_1b3b67077dc5.bed \n",
      "[1] \"nk\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b6c902ae9.bed -b /tmp/RtmpRxRHZx/b_1b3b419b2350.bed \n",
      "[1] \"open\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b334cf5aa.bed -b /tmp/RtmpRxRHZx/b_1b3b671ea359.bed \n",
      "[1] \"progenitor\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b426e53c2.bed -b /tmp/RtmpRxRHZx/b_1b3b1d5ff55d.bed \n",
      "[1] \"T\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b5c542ed3.bed -b /tmp/RtmpRxRHZx/b_1b3b66454f31.bed \n",
      "[1] \"thymocyte\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b273885cd.bed -b /tmp/RtmpRxRHZx/b_1b3b4ec1e3a2.bed \n",
      "[1] \"ncM\"\n",
      "[1] \"gene\"\n",
      "          SNP    gene       beta statistic\n",
      "1 12:56435929   RPS26 -0.5767272 -25.19880\n",
      "2 12:56401085   RPS26 -0.5692010 -23.22648\n",
      "3 12:56470625   RPS26 -0.5417881 -17.47888\n",
      "4 12:10126249 CLEC12A  0.3280741  15.84302\n",
      "5 12:10109648 CLEC12A  0.3151166  15.31361\n",
      "6 12:10109980 CLEC12A  0.3151166  15.31361\n",
      "                                  p.value                               FDR\n",
      "1 0.0000000000000000000000000000001562653 0.0000000000000000000000001733381\n",
      "2 0.0000000000000000000000000000088421238 0.0000000000000000000000049040850\n",
      "3 0.0000000000000000000000068883286751471 0.0000000000000000025469710081834\n",
      "4 0.0000000000000000000005927657381655120 0.0000000000000001023348297632640\n",
      "5 0.0000000000000000000026754083264304801 0.0000000000000001023348297632640\n",
      "6 0.0000000000000000000026754083264304801 0.0000000000000001023348297632640\n",
      "[1] 1109255       6\n",
      "[1] \"expr\"\n",
      "[1] \"no gene\"\n",
      "[1] 1109255       6\n",
      "                SNP    gene                 beta         statistic   p.value\n",
      "1109250 19:36281949 U2AF1L4 -0.00000006842629752 -0.00000889782812 0.9999929\n",
      "1109251 19:36285546 U2AF1L4 -0.00000006842629752 -0.00000889782812 0.9999929\n",
      "1109252 19:36285709 U2AF1L4 -0.00000006842629752 -0.00000889782812 0.9999929\n",
      "1109253  2:63909132    MDH1  0.00000007401522528  0.00000842564641 0.9999933\n",
      "1109254 6:108294145   OSTM1 -0.00000001391555053 -0.00000220832780 0.9999982\n",
      "1109255 19:17885270    JAK3  0.00000000009414278  0.00000002726019 1.0000000\n",
      "              FDR\n",
      "1109250 0.9999951\n",
      "1109251 0.9999951\n",
      "1109252 0.9999951\n",
      "1109253 0.9999951\n",
      "1109254 0.9999991\n",
      "1109255 1.0000000\n",
      "SORTING\n",
      "VALIDATE REGIONS\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      "[1] \"B\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b61518e9b.bed -b /tmp/RtmpRxRHZx/b_1b3b1ccc3906.bed \n",
      "[1] \"EPI\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b6c107885.bed -b /tmp/RtmpRxRHZx/b_1b3b14b857dc.bed \n",
      "[1] \"myeloid\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b6c5a5531.bed -b /tmp/RtmpRxRHZx/b_1b3b576e3b04.bed \n",
      "[1] \"nk\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b7cdd2322.bed -b /tmp/RtmpRxRHZx/b_1b3be593003.bed \n",
      "[1] \"open\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b6dcbaaa.bed -b /tmp/RtmpRxRHZx/b_1b3b6bf0703b.bed \n",
      "[1] \"progenitor\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b129f491b.bed -b /tmp/RtmpRxRHZx/b_1b3b6c9b17a6.bed \n",
      "[1] \"T\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b2db58989.bed -b /tmp/RtmpRxRHZx/b_1b3b313e4f2f.bed \n",
      "[1] \"thymocyte\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b52270a32.bed -b /tmp/RtmpRxRHZx/b_1b3b4f63f408.bed \n",
      "[1] \"T8\"\n",
      "[1] \"gene\"\n",
      "          SNP  gene       beta statistic                                p.value\n",
      "1 12:56435929 RPS26 -0.8688304 -24.00511 0.000000000000000000000000000001739142\n",
      "2 12:56401085 RPS26 -0.8582485 -22.47194 0.000000000000000000000000000044640506\n",
      "3 12:56470625 RPS26 -0.8118072 -16.47246 0.000000000000000000000103053640570197\n",
      "4 12:56384804 RPS26  0.7868531  15.20783 0.000000000000000000003629834738510810\n",
      "5 12:56389293 RPS26  0.7868531  15.20783 0.000000000000000000003629834738510810\n",
      "6 12:56393337 RPS26  0.7868531  15.20783 0.000000000000000000003629834738510810\n",
      "                               FDR\n",
      "1 0.000000000000000000000001707794\n",
      "2 0.000000000000000000000021917930\n",
      "3 0.000000000000000033732032899640\n",
      "4 0.000000000000000594067827891525\n",
      "5 0.000000000000000594067827891525\n",
      "6 0.000000000000000594067827891525\n",
      "[1] 981975      6\n",
      "[1] \"expr\"\n",
      "[1] \"no gene\"\n",
      "[1] 981975      6\n",
      "               SNP    gene              beta      statistic   p.value       FDR\n",
      "981970  7:99719753   COPS6  0.00000010888393  0.00002182502 0.9999827 0.9999871\n",
      "981971 19:40899255 SERTAD1 -0.00000011750652 -0.00002141594 0.9999830 0.9999871\n",
      "981972  6:97362944 NDUFAF4 -0.00000003133775 -0.00001362988 0.9999892 0.9999893\n",
      "981973  6:97363289 NDUFAF4 -0.00000003133775 -0.00001362988 0.9999892 0.9999893\n",
      "981974  6:97369128 NDUFAF4 -0.00000003133775 -0.00001362988 0.9999892 0.9999893\n",
      "981975 12:48154582   RPAP3  0.00000003548726  0.00001346018 0.9999893 0.9999893\n",
      "SORTING\n",
      "VALIDATE REGIONS\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      "[1] \"B\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b1f8f53c5.bed -b /tmp/RtmpRxRHZx/b_1b3b1a055161.bed \n",
      "[1] \"EPI\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b506336f1.bed -b /tmp/RtmpRxRHZx/b_1b3b6812ab1c.bed \n",
      "[1] \"myeloid\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b1c8e584d.bed -b /tmp/RtmpRxRHZx/b_1b3b3d1a7e04.bed \n",
      "[1] \"nk\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b725efb96.bed -b /tmp/RtmpRxRHZx/b_1b3b13fdce5a.bed \n",
      "[1] \"open\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3bd0adc23.bed -b /tmp/RtmpRxRHZx/b_1b3b2f50d0d.bed \n",
      "[1] \"progenitor\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b7386dfb.bed -b /tmp/RtmpRxRHZx/b_1b3b12588994.bed \n",
      "[1] \"T\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b9691567.bed -b /tmp/RtmpRxRHZx/b_1b3b1318ea56.bed \n",
      "[1] \"thymocyte\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b963d6eb.bed -b /tmp/RtmpRxRHZx/b_1b3b2d644d0a.bed \n",
      "[1] \"B\"\n",
      "[1] \"gene\"\n",
      "          SNP  gene       beta statistic                     p.value\n",
      "1 12:56435929 RPS26 -0.8397860 -13.88260 0.0000000000000000001862170\n",
      "2 12:56401085 RPS26 -0.8315262 -13.65659 0.0000000000000000003725119\n",
      "3 22:42475703 SMDT1 -0.1867529 -12.02445 0.0000000000000000675669022\n",
      "4 22:42476754 SMDT1 -0.1867529 -12.02445 0.0000000000000000675669022\n",
      "5 22:42484466 SMDT1 -0.1867529 -12.02445 0.0000000000000000675669022\n",
      "6 22:42487900 SMDT1 -0.1867529 -12.02445 0.0000000000000000675669022\n",
      "                    FDR\n",
      "1 0.0000000000001553742\n",
      "2 0.0000000000001554067\n",
      "3 0.0000000000010995034\n",
      "4 0.0000000000010995034\n",
      "5 0.0000000000010995034\n",
      "6 0.0000000000010995034\n",
      "[1] 834372      6\n",
      "[1] \"expr\"\n",
      "[1] \"no gene\"\n",
      "[1] 834372      6\n",
      "               SNP     gene             beta      statistic   p.value       FDR\n",
      "834367 19:41682274 HNRNPUL1 0.00000006182421 0.000009819184 0.9999922 0.9999938\n",
      "834368 19:48764460     EMP3 0.00000030333637 0.000009354669 0.9999926 0.9999938\n",
      "834369 19:48764721     EMP3 0.00000030333637 0.000009354669 0.9999926 0.9999938\n",
      "834370 19:48764985     EMP3 0.00000030333637 0.000009354669 0.9999926 0.9999938\n",
      "834371 19:48765456     EMP3 0.00000030333637 0.000009354669 0.9999926 0.9999938\n",
      "834372 5:177624342     NHP2 0.00000004161372 0.000004504407 0.9999964 0.9999964\n",
      "SORTING\n",
      "VALIDATE REGIONS\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      "[1] \"B\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b42b21fd2.bed -b /tmp/RtmpRxRHZx/b_1b3b347ddf5a.bed \n",
      "[1] \"EPI\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b475acfbd.bed -b /tmp/RtmpRxRHZx/b_1b3b79c713e5.bed \n",
      "[1] \"myeloid\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b4638a96e.bed -b /tmp/RtmpRxRHZx/b_1b3b630d75e8.bed \n",
      "[1] \"nk\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b67f4ed7e.bed -b /tmp/RtmpRxRHZx/b_1b3b96abd62.bed \n",
      "[1] \"open\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b67fb0b49.bed -b /tmp/RtmpRxRHZx/b_1b3b38ea54c7.bed \n",
      "[1] \"progenitor\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b1dd76ca0.bed -b /tmp/RtmpRxRHZx/b_1b3bfcf609b.bed \n",
      "[1] \"T\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b4bc61a4.bed -b /tmp/RtmpRxRHZx/b_1b3be1d0a9.bed \n",
      "[1] \"thymocyte\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3bf5fc941.bed -b /tmp/RtmpRxRHZx/b_1b3b2746d741.bed \n",
      "[1] \"NK\"\n",
      "[1] \"gene\"\n",
      "          SNP  gene       beta statistic                             p.value\n",
      "1 12:56435929 RPS26 -0.5620326 -20.38888 0.000000000000000000000000004949552\n",
      "2 12:56401085 RPS26 -0.5551429 -19.38220 0.000000000000000000000000055180120\n",
      "3 12:56470625 RPS26 -0.5278729 -15.46050 0.000000000000000000001755310053975\n",
      "4 12:56384804 RPS26  0.5003838  13.01937 0.000000000000000002724393414000780\n",
      "5 12:56389293 RPS26  0.5003838  13.01937 0.000000000000000002724393414000780\n",
      "6 12:56393337 RPS26  0.5003838  13.01937 0.000000000000000002724393414000780\n",
      "                            FDR\n",
      "1 0.000000000000000000004348013\n",
      "2 0.000000000000000000024236930\n",
      "3 0.000000000000000513993400625\n",
      "4 0.000000000000398881164137268\n",
      "5 0.000000000000398881164137268\n",
      "6 0.000000000000398881164137268\n",
      "[1] 878466      6\n",
      "[1] \"expr\"\n",
      "[1] \"no gene\"\n",
      "[1] 878466      6\n",
      "               SNP     gene               beta        statistic   p.value\n",
      "878461 17:74218480   UBALD2  0.000000016696789  0.0000035511308 0.9999972\n",
      "878462 17:74222993   UBALD2  0.000000016696789  0.0000035511308 0.9999972\n",
      "878463 17:74234248   UBALD2  0.000000016696789  0.0000035511308 0.9999972\n",
      "878464 1:168482904     XCL2  0.000000099105518  0.0000033843227 0.9999973\n",
      "878465 1:168483806     XCL2  0.000000099105518  0.0000033843227 0.9999973\n",
      "878466 12:27099677 FGFR1OP2 -0.000000001672715 -0.0000003107173 0.9999998\n",
      "             FDR\n",
      "878461 0.9999985\n",
      "878462 0.9999985\n",
      "878463 0.9999985\n",
      "878464 0.9999985\n",
      "878465 0.9999985\n",
      "878466 0.9999998\n",
      "SORTING\n",
      "VALIDATE REGIONS\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      "[1] \"B\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b680d24.bed -b /tmp/RtmpRxRHZx/b_1b3b3586a095.bed \n",
      "[1] \"EPI\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b4db37ae2.bed -b /tmp/RtmpRxRHZx/b_1b3b28927501.bed \n",
      "[1] \"myeloid\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b37ef4e7b.bed -b /tmp/RtmpRxRHZx/b_1b3bea6f7b3.bed \n",
      "[1] \"nk\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b5f25b4d0.bed -b /tmp/RtmpRxRHZx/b_1b3b75dd1b5b.bed \n",
      "[1] \"open\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b6fa1f7cf.bed -b /tmp/RtmpRxRHZx/b_1b3b515bc1c2.bed \n",
      "[1] \"progenitor\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b3c804916.bed -b /tmp/RtmpRxRHZx/b_1b3bf9fdbb4.bed \n",
      "[1] \"T\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b1ba9c31a.bed -b /tmp/RtmpRxRHZx/b_1b3b458b171f.bed \n",
      "[1] \"thymocyte\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpRxRHZx/a_1b3b7f3bcbc0.bed -b /tmp/RtmpRxRHZx/b_1b3b3da8e206.bed \n"
     ]
    }
   ],
   "source": [
    "for (ct in c('T4', 'cM', 'ncM', 'T8', 'B', 'NK')){\n",
    "print(ct)\n",
    "library(data.table)\n",
    "library(qvalue)\n",
    "library(bedr)\n",
    "options(scipen = 999)\n",
    "# args=commandArgs(TRUE)\n",
    "\n",
    "res_raw=data.frame(fread(paste('/data_volume/memento/lupus/full_analysis/mateqtl/outputs/eur_',ct, '_all_hg19.csv', sep='')))\n",
    "#res_raw=data.frame(fread(args[1]))[-1]\n",
    "#emp=fread(args[2])\n",
    "#delta=fread(args[3])\n",
    "out=paste('/data_volume/memento/lupus/atac_enrichment/gracie/', 'eur_matqetl_',ct, '.out', sep='')\n",
    "#expr_genes=read.table(args[3],header=F)\n",
    "\n",
    "print('gene')\n",
    "#print(head(expr_genes))\n",
    "#print(dim(expr_genes))\n",
    "#print(head(emp))\n",
    "#emp$FDR=qvalue(emp$Empirical)$qvalues\n",
    "#thresh=max(emp$Pvalue[which(emp$FDR < 0.1)])\n",
    "\n",
    "#run if not corrected by celltype\n",
    "#fdr=qvalue(res_raw$pvalue)\n",
    "#print(fdr)\n",
    "#res_raw$FDR=fdr$qvalues\n",
    "\n",
    "#filter FDR\n",
    "#res=res_raw[res_raw$FDR <= 0.1,]\n",
    "#snps gene statistic pvalue FDR beta qvalue\n",
    "\n",
    "\n",
    "#colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\")\n",
    "#take in metasoft data, create atac enrichments from fixed effects\n",
    "colnames(res_raw)=c('SNP', 'gene', 'beta','statistic', 'p.value', 'FDR')\n",
    "# res_raw$gene=sapply(strsplit(res_raw$RSID, '_'), '[', 2)\n",
    "\n",
    "#colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\",\"qvalue\")\n",
    "#res=res_raw\n",
    "#filter for expressed genes\n",
    "print(head(res_raw))\n",
    "print(dim(res_raw))\n",
    "#res_expr=res_raw[res_raw$gene %in% expr_genes$V1,]\n",
    "#print(dim((res_expr)))\n",
    "#res_noexpr=res_raw[!(res_raw$gene %in% expr_genes$V1),]\n",
    "\n",
    "print('expr')\n",
    "#print(dim(res_expr))\n",
    "#print(head(res_expr))\n",
    "print('no gene')\n",
    "#print(head(res_noexpr))\n",
    "\n",
    "#num_no=length(res_noexpr$gene)\n",
    "#print(num_no)\n",
    "#nullp=runif(num_no, min = 0, max = 1)\n",
    "#nullp=rep(1,num_no)\n",
    "#res_noexpr$p.value=nullp\n",
    "#res=rbind(res_expr,res_noexpr)\n",
    "#print(head(res_expr))\n",
    "#print(dim(res_expr))\n",
    "res=res_raw\n",
    "print(dim(res))\n",
    "print(tail(res))\n",
    "#atac=fread('/ye/yelabstore2/10x.lupus/process.scRNAseq/eqtls/S9_lineage_groups.txt.gz')\n",
    "#atac=fread('simple_sorted_atac_lineage_groups2.txt.gz')\n",
    "atac=fread('/data_volume/memento/lupus/atac_enrichment/sorted_simple_atac_lineage_groups3.bed.gz')\n",
    "\n",
    "#names(atac)=c('chr', 'start', 'end', 'group')\n",
    "#atac.bed=data.frame(chr=atac$chr, start=atac$start, end=atac$stop, group=atac$group)\n",
    "atac.groups=names(table(atac$group))\n",
    "#atac.sorted=bedr.sort.region(atac$chr,atac$start,atac$stop)\n",
    "\n",
    "res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "\n",
    "#res$chr=paste('chr', sapply(strsplit(res$snp, ':'), '[', 1), sep='')\n",
    "#res$pos=as.numeric(sapply(strsplit(res$snp, ':'), '[', 2))\n",
    "res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "#res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "#res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "#res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "#res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "all.df=NULL\n",
    "for(g in atac.groups){\n",
    "    print(g)\n",
    "    peaks=atac[which(atac$group == g), ]\n",
    "    peaks=paste(peaks$chr, ':', peaks$start, '-', peaks$stop, sep='')\n",
    "    overlap.sig.snps=bedr(input=list(a=res.bed.sorted, b=peaks), method='intersect')\n",
    "    in_peak=as.numeric(res.bed.sorted %in% overlap.sig.snps)\n",
    "    pvals=res$\"p.value\"[match(res.bed.sorted, res$id)]\n",
    "    #pvals=res$\"pvalue\"[match(res.bed.sorted, res$id)]\n",
    "    #pvals=res$\"p-value\"[match(res.bed.sorted, res$id)]\n",
    "    mw_res=wilcox.test(pvals ~ in_peak, alternative='greater')\n",
    "    df=data.frame(W=mw_res$statistic, pval=mw_res$p.value, group=g)\n",
    "    all.df=rbind(all.df, df)   \n",
    "}\n",
    "\n",
    "write.table(all.df, file=out, row.names=F, col.names=T, quote=F, sep='\\t')\n",
    "\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "92be40cc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
