{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "41517661",
   "metadata": {},
   "source": [
    "### ATAC enrichment"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "a76a263a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "\n",
      "\n",
      "######################\n",
      "#### bedr v1.0.7 ####\n",
      "######################\n",
      "\n",
      "checking binary availability...\n",
      "  * Checking path for bedtools... PASS\n",
      "    /home/ssm-user/anaconda3/envs/single_cell/bin/bedtools\n",
      "  * Checking path for bedops... FAIL\n",
      "  * Checking path for tabix... FAIL\n",
      "tests and examples will be skipped on R CMD check if binaries are missing\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "library(data.table)\n",
    "# library(qvalue)\n",
    "library(bedr)\n",
    "options(scipen = 999)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0c6a08a7",
   "metadata": {},
   "source": [
    "### Copy gracie's attempt with memento  eqtls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b1508ec6",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (ct in c('T4', 'cM', 'ncM', 'T8', 'B', 'NK')){\n",
    "print(paste('starting....',ct))\n",
    "library(data.table)\n",
    "library(qvalue)\n",
    "library(bedr)\n",
    "options(scipen = 999)\n",
    "# args=commandArgs(TRUE)\n",
    "\n",
    "res_raw=data.frame(fread(paste('/data_volume/memento/lupus/full_analysis/memento/100kb/eur_',ct, '.csv', sep='')))\n",
    "#res_raw=data.frame(fread(args[1]))[-1]\n",
    "#emp=fread(args[2])\n",
    "#delta=fread(args[3])\n",
    "out=paste('/data_volume/memento/lupus/atac_enrichment/100kb/', 'eur_memento_',ct, '.out', sep='')\n",
    "#expr_genes=read.table(args[3],header=F)\n",
    "\n",
    "print('gene')\n",
    "#print(head(expr_genes))\n",
    "#print(dim(expr_genes))\n",
    "#print(head(emp))\n",
    "#emp$FDR=qvalue(emp$Empirical)$qvalues\n",
    "#thresh=max(emp$Pvalue[which(emp$FDR < 0.1)])\n",
    "\n",
    "#run if not corrected by celltype\n",
    "#fdr=qvalue(res_raw$pvalue)\n",
    "#print(fdr)\n",
    "#res_raw$FDR=fdr$qvalues\n",
    "\n",
    "#filter FDR\n",
    "#res=res_raw[res_raw$FDR <= 0.1,]\n",
    "#snps gene statistic pvalue FDR beta qvalue\n",
    "\n",
    "\n",
    "#colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\")\n",
    "#take in metasoft data, create atac enrichments from fixed effects\n",
    "colnames(res_raw)=c('SNP', 'gene', 'statistic', 'p.value', 'FDR', 'beta')\n",
    "# res_raw$gene=sapply(strsplit(res_raw$RSID, '_'), '[', 2)\n",
    "\n",
    "#colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\",\"qvalue\")\n",
    "#res=res_raw\n",
    "#filter for expressed genes\n",
    "# print(head(res_raw))\n",
    "# print(dim(res_raw))\n",
    "#res_expr=res_raw[res_raw$gene %in% expr_genes$V1,]\n",
    "#print(dim((res_expr)))\n",
    "#res_noexpr=res_raw[!(res_raw$gene %in% expr_genes$V1),]\n",
    "\n",
    "# print('expr')\n",
    "#print(dim(res_expr))\n",
    "#print(head(res_expr))\n",
    "# print('no gene')\n",
    "#print(head(res_noexpr))\n",
    "\n",
    "#num_no=length(res_noexpr$gene)\n",
    "#print(num_no)\n",
    "#nullp=runif(num_no, min = 0, max = 1)\n",
    "#nullp=rep(1,num_no)\n",
    "#res_noexpr$p.value=nullp\n",
    "#res=rbind(res_expr,res_noexpr)\n",
    "#print(head(res_expr))\n",
    "#print(dim(res_expr))\n",
    "res=res_raw\n",
    "# print(dim(res))\n",
    "# print(tail(res))\n",
    "#atac=fread('/ye/yelabstore2/10x.lupus/process.scRNAseq/eqtls/S9_lineage_groups.txt.gz')\n",
    "#atac=fread('simple_sorted_atac_lineage_groups2.txt.gz')\n",
    "atac=fread('/data_volume/memento/lupus/atac_enrichment/sorted_simple_atac_lineage_groups3.bed.gz')\n",
    "\n",
    "#names(atac)=c('chr', 'start', 'end', 'group')\n",
    "#atac.bed=data.frame(chr=atac$chr, start=atac$start, end=atac$stop, group=atac$group)\n",
    "atac.groups=names(table(atac$group))\n",
    "#atac.sorted=bedr.sort.region(atac$chr,atac$start,atac$stop)\n",
    "\n",
    "res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "\n",
    "#res$chr=paste('chr', sapply(strsplit(res$snp, ':'), '[', 1), sep='')\n",
    "#res$pos=as.numeric(sapply(strsplit(res$snp, ':'), '[', 2))\n",
    "res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "#res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "#res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "#res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "#res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "all.df=NULL\n",
    "for(g in c('B', 'myeloid', 'nk', 'T')){\n",
    "    print(g)\n",
    "    peaks=atac[which(atac$group == g), ]\n",
    "    peaks=paste(peaks$chr, ':', peaks$start, '-', peaks$stop, sep='')\n",
    "    overlap.sig.snps=bedr(input=list(a=res.bed.sorted, b=peaks), method='intersect')\n",
    "    in_peak=as.numeric(res.bed.sorted %in% overlap.sig.snps)\n",
    "    pvals=res$\"p.value\"[match(res.bed.sorted, res$id)]\n",
    "    #pvals=res$\"pvalue\"[match(res.bed.sorted, res$id)]\n",
    "    #pvals=res$\"p-value\"[match(res.bed.sorted, res$id)]\n",
    "    mw_res=wilcox.test(pvals ~ in_peak, alternative='greater')\n",
    "    df=data.frame(W=mw_res$statistic, pval=mw_res$p.value, group=g)\n",
    "    all.df=rbind(all.df, df)   \n",
    "}\n",
    "\n",
    "write.table(all.df, file=out, row.names=F, col.names=T, quote=F, sep='\\t')\n",
    "\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "2acea330",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<style>\n",
       ".list-inline {list-style: none; margin:0; padding: 0}\n",
       ".list-inline>li {display: inline-block}\n",
       ".list-inline>li:not(:last-child)::after {content: \"\\00b7\"; padding: 0 .5ex}\n",
       "</style>\n",
       "<ol class=list-inline><li>'B'</li><li>'EPI'</li><li>'myeloid'</li><li>'nk'</li><li>'open'</li><li>'progenitor'</li><li>'T'</li><li>'thymocyte'</li></ol>\n"
      ],
      "text/latex": [
       "\\begin{enumerate*}\n",
       "\\item 'B'\n",
       "\\item 'EPI'\n",
       "\\item 'myeloid'\n",
       "\\item 'nk'\n",
       "\\item 'open'\n",
       "\\item 'progenitor'\n",
       "\\item 'T'\n",
       "\\item 'thymocyte'\n",
       "\\end{enumerate*}\n"
      ],
      "text/markdown": [
       "1. 'B'\n",
       "2. 'EPI'\n",
       "3. 'myeloid'\n",
       "4. 'nk'\n",
       "5. 'open'\n",
       "6. 'progenitor'\n",
       "7. 'T'\n",
       "8. 'thymocyte'\n",
       "\n",
       "\n"
      ],
      "text/plain": [
       "[1] \"B\"          \"EPI\"        \"myeloid\"    \"nk\"         \"open\"      \n",
       "[6] \"progenitor\" \"T\"          \"thymocyte\" "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "atac.groups"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "2121105e",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<table class=\"dataframe\">\n",
       "<caption>A data.frame: 2 × 5</caption>\n",
       "<thead>\n",
       "\t<tr><th scope=col>W</th><th scope=col>pval</th><th scope=col>conf</th><th scope=col>estimate</th><th scope=col>group</th></tr>\n",
       "\t<tr><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;dbl&gt;</th><th scope=col>&lt;chr&gt;</th></tr>\n",
       "</thead>\n",
       "<tbody>\n",
       "\t<tr><td>4590296557</td><td>0.767464</td><td>-0.006153402</td><td>-0.001645391</td><td>B</td></tr>\n",
       "\t<tr><td>4590296557</td><td>0.767464</td><td>         Inf</td><td>-0.001645391</td><td>B</td></tr>\n",
       "</tbody>\n",
       "</table>\n"
      ],
      "text/latex": [
       "A data.frame: 2 × 5\n",
       "\\begin{tabular}{lllll}\n",
       " W & pval & conf & estimate & group\\\\\n",
       " <dbl> & <dbl> & <dbl> & <dbl> & <chr>\\\\\n",
       "\\hline\n",
       "\t 4590296557 & 0.767464 & -0.006153402 & -0.001645391 & B\\\\\n",
       "\t 4590296557 & 0.767464 &          Inf & -0.001645391 & B\\\\\n",
       "\\end{tabular}\n"
      ],
      "text/markdown": [
       "\n",
       "A data.frame: 2 × 5\n",
       "\n",
       "| W &lt;dbl&gt; | pval &lt;dbl&gt; | conf &lt;dbl&gt; | estimate &lt;dbl&gt; | group &lt;chr&gt; |\n",
       "|---|---|---|---|---|\n",
       "| 4590296557 | 0.767464 | -0.006153402 | -0.001645391 | B |\n",
       "| 4590296557 | 0.767464 |          Inf | -0.001645391 | B |\n",
       "\n"
      ],
      "text/plain": [
       "  W          pval     conf         estimate     group\n",
       "1 4590296557 0.767464 -0.006153402 -0.001645391 B    \n",
       "2 4590296557 0.767464          Inf -0.001645391 B    "
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "conf.int"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3ac4a9c3",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "0d4b7827",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "7ecd0f84",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "252e1dcd",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3bd285f4",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "64b533c8",
   "metadata": {},
   "source": [
    "### Copy gracie's attempt with merged  eqtls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "058dc2da",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (pop in c('asian')){\n",
    "    for (ct in c('T4', 'cM', 'ncM', 'T8', 'B', 'NK')){\n",
    "        print(ct)\n",
    "        library(data.table)\n",
    "        library(qvalue)\n",
    "        library(bedr)\n",
    "        options(scipen = 999)\n",
    "        # args=commandArgs(TRUE)\n",
    "\n",
    "        res_raw=data.frame(fread(paste('/data_volume/memento/lupus/atac_enrichment/merged_eqtls/',pop,'_',ct, '_merged.tsv', sep='')))\n",
    "        #res_raw=data.frame(fread(args[1]))[-1]\n",
    "        #emp=fread(args[2])\n",
    "        #delta=fread(args[3])\n",
    "        out=paste('/data_volume/memento/lupus/atac_enrichment/gracie/', pop, '_',ct, '_merged.out', sep='')\n",
    "        #expr_genes=read.table(args[3],header=F)\n",
    "\n",
    "        print('gene')\n",
    "        #print(head(expr_genes))\n",
    "        #print(dim(expr_genes))\n",
    "        #print(head(emp))\n",
    "        #emp$FDR=qvalue(emp$Empirical)$qvalues\n",
    "        #thresh=max(emp$Pvalue[which(emp$FDR < 0.1)])\n",
    "\n",
    "        #run if not corrected by celltype\n",
    "        #fdr=qvalue(res_raw$pvalue)\n",
    "        #print(fdr)\n",
    "        #res_raw$FDR=fdr$qvalues\n",
    "\n",
    "        #filter FDR\n",
    "        #res=res_raw[res_raw$FDR <= 0.1,]\n",
    "        #snps gene statistic pvalue FDR beta qvalue\n",
    "\n",
    "\n",
    "        #colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\")\n",
    "        #take in metasoft data, create atac enrichments from fixed effects\n",
    "        colnames(res_raw)=c('SNP', 'gene', 'statistic_mem', 'p-value_mem', 'FDR_mem', 'beta_mem', 'beta_meq', 't-stat_meq', 'p-value_meq', 'FDR_meq')\n",
    "        # res_raw$gene=sapply(strsplit(res_raw$RSID, '_'), '[', 2)\n",
    "\n",
    "        #colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\",\"qvalue\")\n",
    "        #res=res_raw\n",
    "        #filter for expressed genes\n",
    "        print(head(res_raw))\n",
    "        print(dim(res_raw))\n",
    "        #res_expr=res_raw[res_raw$gene %in% expr_genes$V1,]\n",
    "        #print(dim((res_expr)))\n",
    "        #res_noexpr=res_raw[!(res_raw$gene %in% expr_genes$V1),]\n",
    "\n",
    "        print('expr')\n",
    "        #print(dim(res_expr))\n",
    "        #print(head(res_expr))\n",
    "        print('no gene')\n",
    "        #print(head(res_noexpr))\n",
    "\n",
    "        #num_no=length(res_noexpr$gene)\n",
    "        #print(num_no)\n",
    "        #nullp=runif(num_no, min = 0, max = 1)\n",
    "        #nullp=rep(1,num_no)\n",
    "        #res_noexpr$p.value=nullp\n",
    "        #res=rbind(res_expr,res_noexpr)\n",
    "        #print(head(res_expr))\n",
    "        #print(dim(res_expr))\n",
    "        res=res_raw\n",
    "        print(dim(res))\n",
    "        print(tail(res))\n",
    "        #atac=fread('/ye/yelabstore2/10x.lupus/process.scRNAseq/eqtls/S9_lineage_groups.txt.gz')\n",
    "        #atac=fread('simple_sorted_atac_lineage_groups2.txt.gz')\n",
    "        atac=fread('/data_volume/memento/lupus/atac_enrichment/sorted_simple_atac_lineage_groups3.bed.gz')\n",
    "\n",
    "        #names(atac)=c('chr', 'start', 'end', 'group')\n",
    "        #atac.bed=data.frame(chr=atac$chr, start=atac$start, end=atac$stop, group=atac$group)\n",
    "        atac.groups=names(table(atac$group))\n",
    "        #atac.sorted=bedr.sort.region(atac$chr,atac$start,atac$stop)\n",
    "\n",
    "        res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "        res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "\n",
    "        #res$chr=paste('chr', sapply(strsplit(res$snp, ':'), '[', 1), sep='')\n",
    "        #res$pos=as.numeric(sapply(strsplit(res$snp, ':'), '[', 2))\n",
    "        res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "        res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "        #res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "        #res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "        #res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "        #res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "        all.df=NULL\n",
    "        for(g in atac.groups){\n",
    "            print(g)\n",
    "            peaks=atac[which(atac$group == g), ]\n",
    "            peaks=paste(peaks$chr, ':', peaks$start, '-', peaks$stop, sep='')\n",
    "            overlap.sig.snps=bedr(input=list(a=res.bed.sorted, b=peaks), method='intersect')\n",
    "            in_peak=as.numeric(res.bed.sorted %in% overlap.sig.snps)\n",
    "            pvals_mem=res$\"p-value_mem\"[match(res.bed.sorted, res$id)]\n",
    "            pvals_meq=res$\"p-value_meq\"[match(res.bed.sorted, res$id)]\n",
    "            #pvals=res$\"pvalue\"[match(res.bed.sorted, res$id)]\n",
    "            #pvals=res$\"p-value\"[match(res.bed.sorted, res$id)]\n",
    "            mw_res_mem=wilcox.test(pvals_mem ~ in_peak, alternative='greater')\n",
    "            mw_res_meq=wilcox.test(pvals_meq ~ in_peak, alternative='greater')\n",
    "\n",
    "            df=data.frame(W_mem=mw_res_mem$statistic, W_meq=mw_res_meq$statistic, pval_mem=mw_res_mem$p.value, pval_meq=mw_res_meq$p.value,group=g)\n",
    "            all.df=rbind(all.df, df)   \n",
    "        }\n",
    "\n",
    "        write.table(all.df, file=out, row.names=F, col.names=T, quote=F, sep='\\t')\n",
    "\n",
    "    }\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ec7a43af",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "0cbbe8a1",
   "metadata": {},
   "source": [
    "### Copy gracie's attempt with mateqtl  eqtls"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "e5fb22d1",
   "metadata": {},
   "outputs": [],
   "source": [
    "for (ct in c('B','T8', 'T4', 'NK', 'ncM', 'cM')){\n",
    "print(ct)\n",
    "library(data.table)\n",
    "library(qvalue)\n",
    "library(bedr)\n",
    "options(scipen = 999)\n",
    "# args=commandArgs(TRUE)\n",
    "\n",
    "res_raw=data.frame(fread(paste('/data_volume/memento/lupus/full_analysis/mateqtl/outputs/eur_',ct, '_all_hg19.csv', sep='')))\n",
    "#res_raw=data.frame(fread(args[1]))[-1]\n",
    "#emp=fread(args[2])\n",
    "#delta=fread(args[3])\n",
    "out=paste('/data_volume/memento/lupus/atac_enrichment/100kb/', 'eur_matqetl_',ct, '.out', sep='')\n",
    "#expr_genes=read.table(args[3],header=F)\n",
    "\n",
    "print('gene')\n",
    "#print(head(expr_genes))\n",
    "#print(dim(expr_genes))\n",
    "#print(head(emp))\n",
    "#emp$FDR=qvalue(emp$Empirical)$qvalues\n",
    "#thresh=max(emp$Pvalue[which(emp$FDR < 0.1)])\n",
    "\n",
    "#run if not corrected by celltype\n",
    "#fdr=qvalue(res_raw$pvalue)\n",
    "#print(fdr)\n",
    "#res_raw$FDR=fdr$qvalues\n",
    "\n",
    "#filter FDR\n",
    "#res=res_raw[res_raw$FDR <= 0.1,]\n",
    "#snps gene statistic pvalue FDR beta qvalue\n",
    "\n",
    "\n",
    "#colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\")\n",
    "#take in metasoft data, create atac enrichments from fixed effects\n",
    "colnames(res_raw)=c('SNP', 'gene', 'beta','statistic', 'p.value', 'FDR')\n",
    "# res_raw$gene=sapply(strsplit(res_raw$RSID, '_'), '[', 2)\n",
    "\n",
    "#colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\",\"qvalue\")\n",
    "#res=res_raw\n",
    "#filter for expressed genes\n",
    "print(head(res_raw))\n",
    "print(dim(res_raw))\n",
    "#res_expr=res_raw[res_raw$gene %in% expr_genes$V1,]\n",
    "#print(dim((res_expr)))\n",
    "#res_noexpr=res_raw[!(res_raw$gene %in% expr_genes$V1),]\n",
    "\n",
    "print('expr')\n",
    "#print(dim(res_expr))\n",
    "#print(head(res_expr))\n",
    "print('no gene')\n",
    "#print(head(res_noexpr))\n",
    "\n",
    "#num_no=length(res_noexpr$gene)\n",
    "#print(num_no)\n",
    "#nullp=runif(num_no, min = 0, max = 1)\n",
    "#nullp=rep(1,num_no)\n",
    "#res_noexpr$p.value=nullp\n",
    "#res=rbind(res_expr,res_noexpr)\n",
    "#print(head(res_expr))\n",
    "#print(dim(res_expr))\n",
    "res=res_raw\n",
    "print(dim(res))\n",
    "print(tail(res))\n",
    "#atac=fread('/ye/yelabstore2/10x.lupus/process.scRNAseq/eqtls/S9_lineage_groups.txt.gz')\n",
    "#atac=fread('simple_sorted_atac_lineage_groups2.txt.gz')\n",
    "atac=fread('/data_volume/memento/lupus/atac_enrichment/sorted_simple_atac_lineage_groups3.bed.gz')\n",
    "\n",
    "#names(atac)=c('chr', 'start', 'end', 'group')\n",
    "#atac.bed=data.frame(chr=atac$chr, start=atac$start, end=atac$stop, group=atac$group)\n",
    "atac.groups=names(table(atac$group))\n",
    "#atac.sorted=bedr.sort.region(atac$chr,atac$start,atac$stop)\n",
    "\n",
    "res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "\n",
    "#res$chr=paste('chr', sapply(strsplit(res$snp, ':'), '[', 1), sep='')\n",
    "#res$pos=as.numeric(sapply(strsplit(res$snp, ':'), '[', 2))\n",
    "res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "#res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "#res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "#res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "#res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "all.df=NULL\n",
    "for(g in atac.groups){\n",
    "    print(g)\n",
    "    peaks=atac[which(atac$group == g), ]\n",
    "    peaks=paste(peaks$chr, ':', peaks$start, '-', peaks$stop, sep='')\n",
    "    overlap.sig.snps=bedr(input=list(a=res.bed.sorted, b=peaks), method='intersect')\n",
    "    in_peak=as.numeric(res.bed.sorted %in% overlap.sig.snps)\n",
    "    pvals=res$\"p.value\"[match(res.bed.sorted, res$id)]\n",
    "    #pvals=res$\"pvalue\"[match(res.bed.sorted, res$id)]\n",
    "    #pvals=res$\"p-value\"[match(res.bed.sorted, res$id)]\n",
    "    mw_res=wilcox.test(pvals ~ in_peak, alternative='greater')\n",
    "    df=data.frame(W=mw_res$statistic, pval=mw_res$p.value, group=g)\n",
    "    all.df=rbind(all.df, df)   \n",
    "}\n",
    "\n",
    "write.table(all.df, file=out, row.names=F, col.names=T, quote=F, sep='\\t')\n",
    "\n",
    "}\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4875846",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "34235354",
   "metadata": {},
   "source": [
    "### Get confidence interval"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "5c782986",
   "metadata": {},
   "outputs": [],
   "source": [
    "matches <- c(\n",
    "    'T4'='T',\n",
    "    'cM'='myeloid',\n",
    "    'ncM'='myeloid',\n",
    "    'B'='B',\n",
    "    'NK'='nk'\n",
    ")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "1a6977e4",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[1] \"T4\"\n",
      "[1] \"gene\"\n",
      "          SNP  gene       beta statistic\n",
      "1 12:56435929 RPS26 -0.9567400 -29.05693\n",
      "2 12:56403577 RPS26  1.0081453  25.35022\n",
      "3 12:56412487 RPS26  1.0081453  25.35022\n",
      "4 12:56416928 RPS26  1.0081453  25.35022\n",
      "5 12:56427808 RPS26  1.0081453  25.35022\n",
      "6 12:56435412 RPS26  0.9745304  25.22708\n",
      "                                        p.value\n",
      "1 0.0000000000000000000000000000000000004792371\n",
      "2 0.0000000000000000000000000000000009385490465\n",
      "3 0.0000000000000000000000000000000009385490465\n",
      "4 0.0000000000000000000000000000000009385490465\n",
      "5 0.0000000000000000000000000000000009385490465\n",
      "6 0.0000000000000000000000000000000012263871276\n",
      "                                      FDR\n",
      "1 0.0000000000000000000000000000008482765\n",
      "2 0.0000000000000000000000000003322568742\n",
      "3 0.0000000000000000000000000003322568742\n",
      "4 0.0000000000000000000000000003322568742\n",
      "5 0.0000000000000000000000000003322568742\n",
      "6 0.0000000000000000000000000003617956489\n",
      "[1] 1770056       6\n",
      "[1] \"expr\"\n",
      "[1] \"no gene\"\n",
      "[1] 1770056       6\n",
      "                SNP    gene                 beta         statistic   p.value\n",
      "1770051 3:182870673   LAMP3 -0.00000000407257085 -0.00000557042509 0.9999956\n",
      "1770052 22:24633621 SPECC1L  0.00000000338456372  0.00000509904891 0.9999959\n",
      "1770053 10:51560207  TIMM23 -0.00000000409051195 -0.00000338276549 0.9999973\n",
      "1770054  4:57908010    NOA1 -0.00000000309142093 -0.00000147258703 0.9999988\n",
      "1770055 19:39316194   SIRT2 -0.00000000170777151 -0.00000066762998 0.9999995\n",
      "1770056 8:124783470 FAM91A1  0.00000000001925019  0.00000001796363 1.0000000\n",
      "              FDR\n",
      "1770051 0.9999982\n",
      "1770052 0.9999982\n",
      "1770053 0.9999990\n",
      "1770054 1.0000000\n",
      "1770055 1.0000000\n",
      "1770056 1.0000000\n",
      "SORTING\n",
      "VALIDATE REGIONS\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      "[1] \"T\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpUY03EU/a_4c575f9ca72d.bed -b /tmp/RtmpUY03EU/b_4c577bd824bb.bed \n",
      "[1] \"cM\"\n",
      "[1] \"gene\"\n",
      "          SNP  gene       beta statistic\n",
      "1 12:56435929 RPS26 -0.5477860 -34.88658\n",
      "2 12:56435412 RPS26  0.5573212  28.37010\n",
      "3 12:56401085 RPS26 -0.5337603 -26.68964\n",
      "4 12:56403577 RPS26  0.5716239  25.50779\n",
      "5 12:56412487 RPS26  0.5716239  25.50779\n",
      "6 12:56416928 RPS26  0.5716239  25.50779\n",
      "                                            p.value\n",
      "1 0.00000000000000000000000000000000000000001510857\n",
      "2 0.00000000000000000000000000000000000182924361515\n",
      "3 0.00000000000000000000000000000000005474682313917\n",
      "4 0.00000000000000000000000000000000066754819772555\n",
      "5 0.00000000000000000000000000000000066754819772555\n",
      "6 0.00000000000000000000000000000000066754819772555\n",
      "                                          FDR\n",
      "1 0.00000000000000000000000000000000002906114\n",
      "2 0.00000000000000000000000000000175926315679\n",
      "3 0.00000000000000000000000000003510160086650\n",
      "4 0.00000000000000000000000000018343146859979\n",
      "5 0.00000000000000000000000000018343146859979\n",
      "6 0.00000000000000000000000000018343146859979\n",
      "[1] 1923487       6\n",
      "[1] \"expr\"\n",
      "[1] \"no gene\"\n",
      "[1] 1923487       6\n",
      "                SNP   gene               beta       statistic   p.value\n",
      "1923482 17:61811642 CCDC47  0.000000008093490  0.000003654403 0.9999971\n",
      "1923483 17:61818390 CCDC47  0.000000008093490  0.000003654403 0.9999971\n",
      "1923484 17:61821209 CCDC47  0.000000008093490  0.000003654403 0.9999971\n",
      "1923485  6:91199220 MAP3K7  0.000000003051664  0.000002652358 0.9999979\n",
      "1923486 19:10131693  EIF3G -0.000000026551222 -0.000002625055 0.9999979\n",
      "1923487  1:66167901  PDE4B -0.000000021118092 -0.000001731219 0.9999986\n",
      "              FDR\n",
      "1923482 0.9999984\n",
      "1923483 0.9999984\n",
      "1923484 0.9999984\n",
      "1923485 0.9999984\n",
      "1923486 0.9999984\n",
      "1923487 0.9999986\n",
      "SORTING\n",
      "VALIDATE REGIONS\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      "[1] \"myeloid\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpUY03EU/a_4c575bf86e02.bed -b /tmp/RtmpUY03EU/b_4c57232141a5.bed \n",
      "[1] \"ncM\"\n",
      "[1] \"gene\"\n",
      "          SNP  gene       beta statistic\n",
      "1 12:56435929 RPS26 -0.5755319 -24.41709\n",
      "2 12:56435412 RPS26  0.5860576  21.89397\n",
      "3 12:56401085 RPS26 -0.5599252 -20.64179\n",
      "4 12:56403577 RPS26  0.6014548  20.53293\n",
      "5 12:56412487 RPS26  0.6014548  20.53293\n",
      "6 12:56416928 RPS26  0.6014548  20.53293\n",
      "                                    p.value                                FDR\n",
      "1 0.000000000000000000000000000000007323848 0.00000000000000000000000001545849\n",
      "2 0.000000000000000000000000000002656476693 0.00000000000000000000000280352065\n",
      "3 0.000000000000000000000000000060519273037 0.00000000000000000000002410972737\n",
      "4 0.000000000000000000000000000079958123774 0.00000000000000000000002410972737\n",
      "5 0.000000000000000000000000000079958123774 0.00000000000000000000002410972737\n",
      "6 0.000000000000000000000000000079958123774 0.00000000000000000000002410972737\n",
      "[1] 2110706       6\n",
      "[1] \"expr\"\n",
      "[1] \"no gene\"\n",
      "[1] 2110706       6\n",
      "                SNP      gene               beta       statistic   p.value\n",
      "2110701  8:22800392 TNFRSF10B  0.000000020124537  0.000006711606 0.9999947\n",
      "2110702  8:22800533 TNFRSF10B  0.000000020124537  0.000006711606 0.9999947\n",
      "2110703 16:57055414   HERPUD1  0.000000111681752  0.000004809824 0.9999962\n",
      "2110704 16:57056507   HERPUD1  0.000000111681752  0.000004809824 0.9999962\n",
      "2110705 2:135609247     CCNT2  0.000000008094055  0.000002136140 0.9999983\n",
      "2110706 3:114126247    ZBTB20 -0.000000004515320 -0.000000979320 0.9999992\n",
      "              FDR\n",
      "2110701 0.9999966\n",
      "2110702 0.9999966\n",
      "2110703 0.9999971\n",
      "2110704 0.9999971\n",
      "2110705 0.9999988\n",
      "2110706 0.9999992\n",
      "SORTING\n",
      "VALIDATE REGIONS\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      "[1] \"myeloid\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpUY03EU/a_4c57acb497b.bed -b /tmp/RtmpUY03EU/b_4c572f664ecc.bed \n",
      "[1] \"T8\"\n",
      "[1] \"gene\"\n",
      "          SNP  gene       beta statistic\n",
      "1 12:56435929 RPS26 -0.8903262 -24.60162\n",
      "2 12:56435412 RPS26  0.9120478  23.32615\n",
      "3 12:56403577 RPS26  0.9430783  23.31739\n",
      "4 12:56412487 RPS26  0.9430783  23.31739\n",
      "5 12:56416928 RPS26  0.9430783  23.31739\n",
      "6 12:56427808 RPS26  0.9430783  23.31739\n",
      "                                    p.value                                 FDR\n",
      "1 0.000000000000000000000000000000004853745 0.000000000000000000000000008789288\n",
      "2 0.000000000000000000000000000000087916923 0.000000000000000000000000027078728\n",
      "3 0.000000000000000000000000000000089722794 0.000000000000000000000000027078728\n",
      "4 0.000000000000000000000000000000089722794 0.000000000000000000000000027078728\n",
      "5 0.000000000000000000000000000000089722794 0.000000000000000000000000027078728\n",
      "6 0.000000000000000000000000000000089722794 0.000000000000000000000000027078728\n",
      "[1] 1810826       6\n",
      "[1] \"expr\"\n",
      "[1] \"no gene\"\n",
      "[1] 1810826       6\n",
      "                SNP    gene                beta       statistic   p.value\n",
      "1810821 13:32983250 N4BP2L1 -0.0000000401949457 -0.000010402220 0.9999917\n",
      "1810822 7:138074809  TRIM24  0.0000000030562590  0.000003123445 0.9999975\n",
      "1810823 9:117628528  TNFSF8  0.0000000038887757  0.000001678499 0.9999987\n",
      "1810824 9:117629709  TNFSF8  0.0000000038887757  0.000001678499 0.9999987\n",
      "1810825 9:117630043  TNFSF8  0.0000000038887757  0.000001678499 0.9999987\n",
      "1810826  8:37659282  ERLIN2  0.0000000007382393  0.000001107666 0.9999991\n",
      "              FDR\n",
      "1810821 0.9999945\n",
      "1810822 0.9999991\n",
      "1810823 0.9999991\n",
      "1810824 0.9999991\n",
      "1810825 0.9999991\n",
      "1810826 0.9999991\n",
      "SORTING\n",
      "VALIDATE REGIONS\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      "[1] \"T\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpUY03EU/a_4c57353ce640.bed -b /tmp/RtmpUY03EU/b_4c57a0fdfc7.bed \n",
      "[1] \"nk\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpUY03EU/a_4c5716503679.bed -b /tmp/RtmpUY03EU/b_4c577248a47c.bed \n",
      "[1] \"B\"\n",
      "[1] \"gene\"\n",
      "          SNP  gene       beta statistic\n",
      "1 12:56435929 RPS26 -0.8653580 -28.42135\n",
      "2 12:56401085 RPS26 -0.8479756 -24.79729\n",
      "3 12:56403577 RPS26  0.9088036  24.01702\n",
      "4 12:56412487 RPS26  0.9088036  24.01702\n",
      "5 12:56416928 RPS26  0.9088036  24.01702\n",
      "6 12:56427808 RPS26  0.9088036  24.01702\n",
      "                                       p.value\n",
      "1 0.000000000000000000000000000000000001653668\n",
      "2 0.000000000000000000000000000000003146547902\n",
      "3 0.000000000000000000000000000000018026583938\n",
      "4 0.000000000000000000000000000000018026583938\n",
      "5 0.000000000000000000000000000000018026583938\n",
      "6 0.000000000000000000000000000000018026583938\n",
      "                                   FDR\n",
      "1 0.0000000000000000000000000000029996\n",
      "2 0.0000000000000000000000000028537726\n",
      "3 0.0000000000000000000000000054497578\n",
      "4 0.0000000000000000000000000054497578\n",
      "5 0.0000000000000000000000000054497578\n",
      "6 0.0000000000000000000000000054497578\n",
      "[1] 1813907       6\n",
      "[1] \"expr\"\n",
      "[1] \"no gene\"\n",
      "[1] 1813907       6\n",
      "               SNP   gene               beta       statistic   p.value\n",
      "1813902 1:70636364 LRRC40  0.000000004944438  0.000003823513 0.9999970\n",
      "1813903 1:70647802 LRRC40  0.000000004944438  0.000003823513 0.9999970\n",
      "1813904 1:70649135 LRRC40  0.000000004944438  0.000003823513 0.9999970\n",
      "1813905 2:98169742  COX5B -0.000000050849935 -0.000002655489 0.9999979\n",
      "1813906 2:98170123  COX5B -0.000000050849935 -0.000002655489 0.9999979\n",
      "1813907 2:98170267  COX5B -0.000000050849935 -0.000002655489 0.9999979\n",
      "              FDR\n",
      "1813902 0.9999979\n",
      "1813903 0.9999979\n",
      "1813904 0.9999979\n",
      "1813905 0.9999979\n",
      "1813906 0.9999979\n",
      "1813907 0.9999979\n",
      "SORTING\n",
      "VALIDATE REGIONS\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      "[1] \"B\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpUY03EU/a_4c57796e75eb.bed -b /tmp/RtmpUY03EU/b_4c57439bf67.bed \n",
      "[1] \"NK\"\n",
      "[1] \"gene\"\n",
      "          SNP  gene       beta statistic\n",
      "1 12:56435929 RPS26 -0.5808788 -23.23675\n",
      "2 12:56401085 RPS26 -0.5691345 -21.04113\n",
      "3 12:56435412 RPS26  0.5873350  19.86378\n",
      "4 12:56403577 RPS26  0.6037294  19.01819\n",
      "5 12:56412487 RPS26  0.6037294  19.01819\n",
      "6 12:56416928 RPS26  0.6037294  19.01819\n",
      "                                  p.value                               FDR\n",
      "1 0.0000000000000000000000000000001082239 0.0000000000000000000000002072826\n",
      "2 0.0000000000000000000000000000219897339 0.0000000000000000000000210586116\n",
      "3 0.0000000000000000000000000004542100751 0.0000000000000000000002899848205\n",
      "4 0.0000000000000000000000000043443198008 0.0000000000000000000011886760272\n",
      "5 0.0000000000000000000000000043443198008 0.0000000000000000000011886760272\n",
      "6 0.0000000000000000000000000043443198008 0.0000000000000000000011886760272\n",
      "[1] 1915313       6\n",
      "[1] \"expr\"\n",
      "[1] \"no gene\"\n",
      "[1] 1915313       6\n",
      "               SNP   gene                beta         statistic   p.value\n",
      "1915308 7:45054015  MYO1G -0.0000000649226235 -0.00000475676808 0.9999962\n",
      "1915309 1:89269340  GTF2B -0.0000000093116716 -0.00000164289983 0.9999987\n",
      "1915310  1:1708801 CDK11A -0.0000000004029278 -0.00000003849408 1.0000000\n",
      "1915311  1:1721479 CDK11A -0.0000000004029278 -0.00000003849408 1.0000000\n",
      "1915312  1:1722932 CDK11A -0.0000000004029278 -0.00000003849408 1.0000000\n",
      "1915313  1:1723031 CDK11A -0.0000000004029278 -0.00000003849408 1.0000000\n",
      "              FDR\n",
      "1915308 0.9999988\n",
      "1915309 1.0000000\n",
      "1915310 1.0000000\n",
      "1915311 1.0000000\n",
      "1915312 1.0000000\n",
      "1915313 1.0000000\n",
      "SORTING\n",
      "VALIDATE REGIONS\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      "[1] \"nk\"\n",
      " * Processing input (1): a\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... FAIL\n",
      "   The input for object has overlapping features!\n",
      "   This can cause unexpected results for some set operations.\n",
      "   i.e. x <- bedr.merge.region(x)\n",
      " * Processing input (2): b\n",
      "CONVERT TO BED\n",
      " * Checking input type... PASS\n",
      "   Input is in index format\n",
      "VALIDATE REGIONS\n",
      " * Check if index is a string... PASS\n",
      " * Check index pattern... PASS\n",
      " * Check for missing values... PASS\n",
      " * Check for larger start position... PASS.\n",
      " * Check if zero based... PASS\n",
      " * Checking sort order... PASS\n",
      " * Checking for overlapping 'contiguous' regions... PASS\n",
      "   bedtools intersect -a /tmp/RtmpUY03EU/a_4c5714e3325e.bed -b /tmp/RtmpUY03EU/b_4c574a20189e.bed \n"
     ]
    }
   ],
   "source": [
    "all.df=NULL\n",
    "pop <- 'asian'\n",
    "for (ct in c('T4', 'cM', 'ncM', 'T8', 'B', 'NK')){\n",
    "    print(ct)\n",
    "    library(data.table)\n",
    "    library(qvalue)\n",
    "    library(bedr)\n",
    "    options(scipen = 999)\n",
    "    # args=commandArgs(TRUE)\n",
    "\n",
    "#     res_raw=data.frame(fread(paste('/data_volume/memento/lupus/full_analysis/memento/100kb/', pop, '_',ct, '.csv', sep='')))\n",
    "    res_raw=data.frame(fread(paste('/data_volume/memento/lupus/full_analysis/mateqtl/outputs/', pop, '_',ct, '_all_hg19.csv', sep='')))\n",
    "\n",
    "    #res_raw=data.frame(fread(args[1]))[-1]\n",
    "    #emp=fread(args[2])\n",
    "    #delta=fread(args[3])\n",
    "    #expr_genes=read.table(args[3],header=F)\n",
    "\n",
    "    print('gene')\n",
    "    #print(head(expr_genes))\n",
    "    #print(dim(expr_genes))\n",
    "    #print(head(emp))\n",
    "    #emp$FDR=qvalue(emp$Empirical)$qvalues\n",
    "    #thresh=max(emp$Pvalue[which(emp$FDR < 0.1)])\n",
    "\n",
    "    #run if not corrected by celltype\n",
    "    #fdr=qvalue(res_raw$pvalue)\n",
    "    #print(fdr)\n",
    "    #res_raw$FDR=fdr$qvalues\n",
    "\n",
    "    #filter FDR\n",
    "    #res=res_raw[res_raw$FDR <= 0.1,]\n",
    "    #snps gene statistic pvalue FDR beta qvalue\n",
    "\n",
    "\n",
    "    #colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\")\n",
    "    #take in metasoft data, create atac enrichments from fixed effects\n",
    "    colnames(res_raw)=c('SNP', 'gene', 'beta','statistic', 'p.value', 'FDR')\n",
    "    # res_raw$gene=sapply(strsplit(res_raw$RSID, '_'), '[', 2)\n",
    "\n",
    "    #colnames(res_raw)=c(\"SNP\",\"gene\",\"statistic\",\"p.value\",\"FDR\",\"beta\",\"qvalue\")\n",
    "    #res=res_raw\n",
    "    #filter for expressed genes\n",
    "    print(head(res_raw))\n",
    "    print(dim(res_raw))\n",
    "    #res_expr=res_raw[res_raw$gene %in% expr_genes$V1,]\n",
    "    #print(dim((res_expr)))\n",
    "    #res_noexpr=res_raw[!(res_raw$gene %in% expr_genes$V1),]\n",
    "\n",
    "    print('expr')\n",
    "    #print(dim(res_expr))\n",
    "    #print(head(res_expr))\n",
    "    print('no gene')\n",
    "    #print(head(res_noexpr))\n",
    "\n",
    "    #num_no=length(res_noexpr$gene)\n",
    "    #print(num_no)\n",
    "    #nullp=runif(num_no, min = 0, max = 1)\n",
    "    #nullp=rep(1,num_no)\n",
    "    #res_noexpr$p.value=nullp\n",
    "    #res=rbind(res_expr,res_noexpr)\n",
    "    #print(head(res_expr))\n",
    "    #print(dim(res_expr))\n",
    "    res=res_raw\n",
    "    print(dim(res))\n",
    "    print(tail(res))\n",
    "    #atac=fread('/ye/yelabstore2/10x.lupus/process.scRNAseq/eqtls/S9_lineage_groups.txt.gz')\n",
    "    #atac=fread('simple_sorted_atac_lineage_groups2.txt.gz')\n",
    "    atac=fread('/data_volume/memento/lupus/atac_enrichment/sorted_simple_atac_lineage_groups3.bed.gz')\n",
    "\n",
    "    #names(atac)=c('chr', 'start', 'end', 'group')\n",
    "    #atac.bed=data.frame(chr=atac$chr, start=atac$start, end=atac$stop, group=atac$group)\n",
    "    atac.groups=names(table(atac$group))\n",
    "    #atac.sorted=bedr.sort.region(atac$chr,atac$start,atac$stop)\n",
    "\n",
    "    res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "    res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "\n",
    "    #res$chr=paste('chr', sapply(strsplit(res$snp, ':'), '[', 1), sep='')\n",
    "    #res$pos=as.numeric(sapply(strsplit(res$snp, ':'), '[', 2))\n",
    "    res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "    res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "\n",
    "    #res$chr=paste('chr', sapply(strsplit(res$SNP, ':'), '[', 1), sep='')\n",
    "    #res$pos=as.numeric(sapply(strsplit(res$SNP, ':'), '[', 2))\n",
    "    #res$id=paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')\n",
    "    #res.bed.sorted=bedr.sort.region(unique(paste(res$chr, ':', res$pos -1, '-' ,res$pos, sep='')))\n",
    "    \n",
    "    if (ct == 'T8'){\n",
    "        groups <- c('T', 'nk')\n",
    "    }\n",
    "    else {\n",
    "        groups <- c(matches[ct])\n",
    "    }\n",
    "\n",
    "    for(g in groups){\n",
    "        print(g)\n",
    "        peaks=atac[which(atac$group == g), ]\n",
    "        peaks=paste(peaks$chr, ':', peaks$start, '-', peaks$stop, sep='')\n",
    "        overlap.sig.snps=bedr(input=list(a=res.bed.sorted, b=peaks), method='intersect')\n",
    "        in_peak=as.numeric(res.bed.sorted %in% overlap.sig.snps)\n",
    "        pvals=res$\"p.value\"[match(res.bed.sorted, res$id)]\n",
    "        write.table(\n",
    "            cbind(in_peak, pvals),\n",
    "            file=paste('/data_volume/memento/lupus/atac_enrichment/matched/mateqtl/', pop, '_',ct, '_', g,'.txt', sep=''), \n",
    "            row.names=F, col.names=T, quote=F, sep='\\t')\n",
    "        \n",
    "    }\n",
    "    \n",
    "\n",
    "}\n",
    "# write.table(all.df, file='asian_memento_confint.txt', row.names=F, col.names=T, quote=F, sep='\\t')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "bd1ac0da",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
