{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "c70b0eb5",
   "metadata": {},
   "source": [
    "# Matrix eQTL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e138c844",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"MatrixEQTL\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "68ba244e",
   "metadata": {},
   "outputs": [],
   "source": [
    "base.dir = '/data_volume/memento/lupus/mateqtl_input/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "3ae6b25e",
   "metadata": {},
   "outputs": [],
   "source": [
    "useModel = modelLINEAR; # modelANOVA, modelLINEAR, or modelLINEAR_CROSS\n",
    "\n",
    "# Genotype file name\n",
    "SNP_file_name = paste(base.dir, \"filtered_genos.tsv\", sep=\"\");\n",
    "\n",
    "# Covariates file name\n",
    "# Set to character() for no covariates\n",
    "covariates_file_name = paste(base.dir, \"cm_cg.mateqtl_cov.txt\", sep=\"\");\n",
    "\n",
    "# Output file name\n",
    "# output_file_name_cis = paste(base.dir, \"cm_cg_matqtl_cis.out\", sep=\"\");\n",
    "\n",
    "# Only associations significant at this level will be saved\n",
    "pvOutputThreshold = 1;\n",
    "# pvOutputThreshold_tra = 0;\n",
    "\n",
    "\n",
    "# Error covariance matrix\n",
    "# Set to numeric() for identity.\n",
    "errorCovariance = numeric();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "5059922d",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Rows read: 7917 done.\n",
      "\n",
      "Rows read: 21 done.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## Load genotype data\n",
    "\n",
    "snps = SlicedData$new();\n",
    "snps$fileDelimiter = \"\\t\";      # the TAB character\n",
    "snps$fileOmitCharacters = \"NA\"; # denote missing values;\n",
    "snps$fileSkipRows = 1;          # one row of column labels\n",
    "snps$fileSkipColumns = 1;       # one column of row labels\n",
    "snps$fileSliceSize = 100000;      # read file in slices of 2,000 rows\n",
    "snps$LoadFile(SNP_file_name);\n",
    "\n",
    "\n",
    "## Load covariates\n",
    "\n",
    "cvrt = SlicedData$new();\n",
    "cvrt$fileDelimiter = \"\\t\";      # the TAB character\n",
    "cvrt$fileOmitCharacters = \"NA\"; # denote missing values;\n",
    "cvrt$fileSkipRows = 1;          # one row of column labels\n",
    "cvrt$fileSkipColumns = 1;       # one column of row labels\n",
    "if(length(covariates_file_name)>0) {\n",
    "cvrt$LoadFile(covariates_file_name);\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "dded1a6a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# snps_location_file_name = paste(base.dir, \"cm_cg.snpspos.txt\", sep=\"\");\n",
    "# gene_location_file_name = paste(base.dir, \"../geneloc.txt\", sep=\"\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d2ad44f1",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cisDist = 1e5;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "bfebebd8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# snpspos = read.table(snps_location_file_name, header = TRUE, stringsAsFactors = FALSE);\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "35a9bf8c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genepos = read.table(gene_location_file_name, header = TRUE, stringsAsFactors = FALSE, sep=\",\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "c6d9b42a",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Rows read: 502 done.\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.003 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 0.04 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Warning message in Matrix_eQTL_main(snps = snps, gene = gene, cvrt = cvrt, output_file_name = output_file_name, :\n",
      "“Warning: pvOutputThreshold may be too large.\n",
      "Expected number of findings > 3974334”\n",
      "Task finished in 0.097 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "100.00% done, 3,974,334 eQTLs\n",
      "\n",
      "Task finished in 21.541 seconds\n",
      "\n",
      "\n",
      "\n",
      "Rows read: 836 done.\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.002 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 0.055 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Warning message in Matrix_eQTL_main(snps = snps, gene = gene, cvrt = cvrt, output_file_name = output_file_name, :\n",
      "“Warning: pvOutputThreshold may be too large.\n",
      "Expected number of findings > 6618612”\n",
      "Task finished in 0.072 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "100.00% done, 6,618,612 eQTLs\n",
      "\n",
      "Task finished in 34.779 seconds\n",
      "\n",
      "\n",
      "\n",
      "Rows read: 966 done.\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.002 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 0.013 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Warning message in Matrix_eQTL_main(snps = snps, gene = gene, cvrt = cvrt, output_file_name = output_file_name, :\n",
      "“Warning: pvOutputThreshold may be too large.\n",
      "Expected number of findings > 7647822”\n",
      "Task finished in 0.09 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "100.00% done, 7,647,822 eQTLs\n",
      "\n",
      "Task finished in 40.111 seconds\n",
      "\n",
      "\n",
      "\n",
      "Rows read: 466 done.\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.004 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 0.036 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Warning message in Matrix_eQTL_main(snps = snps, gene = gene, cvrt = cvrt, output_file_name = output_file_name, :\n",
      "“Warning: pvOutputThreshold may be too large.\n",
      "Expected number of findings > 3689322”\n",
      "Task finished in 0.053 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "100.00% done, 3,689,322 eQTLs\n",
      "\n",
      "Task finished in 19.451 seconds\n",
      "\n",
      "\n",
      "\n",
      "Rows read: 133 done.\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.002 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 0.008 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Warning message in Matrix_eQTL_main(snps = snps, gene = gene, cvrt = cvrt, output_file_name = output_file_name, :\n",
      "“Warning: pvOutputThreshold may be too large.\n",
      "Expected number of findings > 1052961”\n",
      "Task finished in 0.03 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "100.00% done, 1,052,961 eQTLs\n",
      "\n",
      "Task finished in 5.679 seconds\n",
      "\n",
      "\n",
      "\n",
      "Rows read: 258 done.\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.002 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 0.01 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Warning message in Matrix_eQTL_main(snps = snps, gene = gene, cvrt = cvrt, output_file_name = output_file_name, :\n",
      "“Warning: pvOutputThreshold may be too large.\n",
      "Expected number of findings > 2042586”\n",
      "Task finished in 0.04 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "100.00% done, 2,042,586 eQTLs\n",
      "\n",
      "Task finished in 10.782 seconds\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for (ct in c('B', 'T8', 'T4', 'NK', 'ncM', 'cM')){\n",
    "    \n",
    "#     ct <- 'cM'\n",
    "    # Gene expression file name\n",
    "    expression_file_name = paste(base.dir, \"../pseudobulk/\", ct,'_filtered.csv',sep=\"\");\n",
    "#     expression_file_name = paste(base.dir, \"cm_cg.expr\",sep=\"\")\n",
    "    output_file_name = paste(base.dir, \"../mateqtl_output/\",ct, \"_filtered.out\", sep=\"\");\n",
    "\n",
    "\n",
    "    ## Load gene expression data\n",
    "\n",
    "    gene = SlicedData$new();\n",
    "    gene$fileDelimiter = \"\\t\";      # the TAB character\n",
    "    gene$fileOmitCharacters = \"NA\"; # denote missing values;\n",
    "    gene$fileSkipRows = 1;          # one row of column labels\n",
    "    gene$fileSkipColumns = 1;       # one column of row labels\n",
    "    gene$fileSliceSize = 2000;      # read file in slices of 2,000 rows\n",
    "    gene$LoadFile(expression_file_name);\n",
    "\n",
    "\n",
    "    me = Matrix_eQTL_main(\n",
    "        snps = snps,\n",
    "        gene = gene,\n",
    "        cvrt = cvrt,\n",
    "        output_file_name     = output_file_name,\n",
    "        pvOutputThreshold     = pvOutputThreshold,\n",
    "        useModel = useModel,\n",
    "        errorCovariance = errorCovariance,\n",
    "        verbose = TRUE,\n",
    "    #     output_file_name.cis = output_file_name_cis,\n",
    "    #     pvOutputThreshold.cis = pvOutputThreshold_cis,\n",
    "    #     snpspos = snpspos,\n",
    "    #     genepos = genepos,\n",
    "    #     cisDist = cisDist,\n",
    "        pvalue.hist = \"qqplot\",\n",
    "        min.pv.by.genesnp = FALSE,\n",
    "        noFDRsaveMemory = FALSE);\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "8c3c8863",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ca03faa9",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a4d83b3e",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a9d1a1f6",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3e055482",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "5e8ae62b",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
