{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "20aa782d",
   "metadata": {},
   "source": [
    "# Matrix eQTL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "d0b139e3",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"MatrixEQTL\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "5ee01585",
   "metadata": {},
   "outputs": [],
   "source": [
    "base.dir = '/data_volume/memento/lupus/mateqtl_input/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "f21f48ac",
   "metadata": {},
   "outputs": [],
   "source": [
    "useModel = modelLINEAR; # modelANOVA, modelLINEAR, or modelLINEAR_CROSS\n",
    "\n",
    "# Genotype file name\n",
    "SNP_file_name = paste(base.dir, \"filtered_genos.tsv\", sep=\"\");\n",
    "\n",
    "# Covariates file name\n",
    "# Set to character() for no covariates\n",
    "covariates_file_name = paste(base.dir, \"mateqtl_cov.txt\", sep=\"\");\n",
    "\n",
    "# Output file name\n",
    "# output_file_name_cis = paste(base.dir, \"cm_cg_matqtl_cis.out\", sep=\"\");\n",
    "\n",
    "# Only associations significant at this level will be saved\n",
    "pvOutputThreshold = 1;\n",
    "# pvOutputThreshold_tra = 0;\n",
    "\n",
    "\n",
    "# Error covariance matrix\n",
    "# Set to numeric() for identity.\n",
    "errorCovariance = numeric();"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "a77ac015",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Rows read: 9630 done.\n",
      "\n",
      "Rows read: 35 done.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "## Load genotype data\n",
    "\n",
    "snps = SlicedData$new();\n",
    "snps$fileDelimiter = \"\\t\";      # the TAB character\n",
    "snps$fileOmitCharacters = \"NA\"; # denote missing values;\n",
    "snps$fileSkipRows = 1;          # one row of column labels\n",
    "snps$fileSkipColumns = 1;       # one column of row labels\n",
    "snps$fileSliceSize = 100000;      # read file in slices of 2,000 rows\n",
    "snps$LoadFile(SNP_file_name);\n",
    "\n",
    "\n",
    "## Load covariates\n",
    "\n",
    "cvrt = SlicedData$new();\n",
    "cvrt$fileDelimiter = \"\\t\";      # the TAB character\n",
    "cvrt$fileOmitCharacters = \"NA\"; # denote missing values;\n",
    "cvrt$fileSkipRows = 1;          # one row of column labels\n",
    "cvrt$fileSkipColumns = 1;       # one column of row labels\n",
    "if(length(covariates_file_name)>0) {\n",
    "cvrt$LoadFile(covariates_file_name);\n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "c6797e61",
   "metadata": {},
   "outputs": [],
   "source": [
    "# snps_location_file_name = paste(base.dir, \"cm_cg.snpspos.txt\", sep=\"\");\n",
    "# gene_location_file_name = paste(base.dir, \"../geneloc.txt\", sep=\"\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "8f74d104",
   "metadata": {},
   "outputs": [],
   "source": [
    "# cisDist = 1e5;\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "1b546b84",
   "metadata": {},
   "outputs": [],
   "source": [
    "# snpspos = read.table(snps_location_file_name, header = TRUE, stringsAsFactors = FALSE);\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "c107cec6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# genepos = read.table(gene_location_file_name, header = TRUE, stringsAsFactors = FALSE, sep=\",\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "6786998f",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Rows read: 609 done.\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.003 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 0.024 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Warning message in Matrix_eQTL_main(snps = snps, gene = gene, cvrt = cvrt, output_file_name = output_file_name, :\n",
      "“Warning: pvOutputThreshold may be too large.\n",
      "Expected number of findings > 5864670”\n",
      "Task finished in 0.083 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "100.00% done, 5,864,670 eQTLs\n",
      "\n",
      "Task finished in 31.682 seconds\n",
      "\n",
      "\n",
      "\n",
      "Rows read: 1085 done.\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.002 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 0.037 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Warning message in Matrix_eQTL_main(snps = snps, gene = gene, cvrt = cvrt, output_file_name = output_file_name, :\n",
      "“Warning: pvOutputThreshold may be too large.\n",
      "Expected number of findings > 10448550”\n",
      "Task finished in 0.084 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "100.00% done, 10,448,550 eQTLs\n",
      "\n",
      "Task finished in 55.56 seconds\n",
      "\n",
      "\n",
      "\n",
      "Rows read: 1315 done.\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.002 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 0.041 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Warning message in Matrix_eQTL_main(snps = snps, gene = gene, cvrt = cvrt, output_file_name = output_file_name, :\n",
      "“Warning: pvOutputThreshold may be too large.\n",
      "Expected number of findings > 12663450”\n",
      "Task finished in 0.112 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "100.00% done, 12,663,450 eQTLs\n",
      "\n",
      "Task finished in 66.243 seconds\n",
      "\n",
      "\n",
      "\n",
      "Rows read: 763 done.\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.002 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 0.035 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Warning message in Matrix_eQTL_main(snps = snps, gene = gene, cvrt = cvrt, output_file_name = output_file_name, :\n",
      "“Warning: pvOutputThreshold may be too large.\n",
      "Expected number of findings > 7347690”\n",
      "Task finished in 0.08 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "100.00% done, 7,347,690 eQTLs\n",
      "\n",
      "Task finished in 38.589 seconds\n",
      "\n",
      "\n",
      "\n",
      "Rows read: 230 done.\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.003 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 0.03 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Warning message in Matrix_eQTL_main(snps = snps, gene = gene, cvrt = cvrt, output_file_name = output_file_name, :\n",
      "“Warning: pvOutputThreshold may be too large.\n",
      "Expected number of findings > 2214900”\n",
      "Task finished in 0.037 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "100.00% done, 2,214,900 eQTLs\n",
      "\n",
      "Task finished in 11.702 seconds\n",
      "\n",
      "\n",
      "\n",
      "Rows read: 340 done.\n",
      "\n",
      "Processing covariates\n",
      "\n",
      "Task finished in 0.002 seconds\n",
      "\n",
      "Processing gene expression data (imputation, residualization)\n",
      "\n",
      "Task finished in 0.007 seconds\n",
      "\n",
      "Creating output file(s)\n",
      "\n",
      "Warning message in Matrix_eQTL_main(snps = snps, gene = gene, cvrt = cvrt, output_file_name = output_file_name, :\n",
      "“Warning: pvOutputThreshold may be too large.\n",
      "Expected number of findings > 3274200”\n",
      "Task finished in 0.052 seconds\n",
      "\n",
      "Performing eQTL analysis\n",
      "\n",
      "100.00% done, 3,274,200 eQTLs\n",
      "\n",
      "Task finished in 16.998 seconds\n",
      "\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for (ct in c('B', 'T8', 'T4', 'NK', 'ncM', 'cM')){\n",
    "    \n",
    "#     ct <- 'cM'\n",
    "    # Gene expression file name\n",
    "    expression_file_name = paste(base.dir, \"../pseudobulk/\", ct,'_filtered.csv',sep=\"\");\n",
    "#     expression_file_name = paste(base.dir, \"cm_cg.expr\",sep=\"\")\n",
    "    output_file_name = paste(base.dir, \"../mateqtl_output/\",ct, \"_filtered.out\", sep=\"\");\n",
    "\n",
    "\n",
    "    ## Load gene expression data\n",
    "\n",
    "    gene = SlicedData$new();\n",
    "    gene$fileDelimiter = \"\\t\";      # the TAB character\n",
    "    gene$fileOmitCharacters = \"NA\"; # denote missing values;\n",
    "    gene$fileSkipRows = 1;          # one row of column labels\n",
    "    gene$fileSkipColumns = 1;       # one column of row labels\n",
    "    gene$fileSliceSize = 2000;      # read file in slices of 2,000 rows\n",
    "    gene$LoadFile(expression_file_name);\n",
    "\n",
    "\n",
    "    me = Matrix_eQTL_main(\n",
    "        snps = snps,\n",
    "        gene = gene,\n",
    "        cvrt = cvrt,\n",
    "        output_file_name     = output_file_name,\n",
    "        pvOutputThreshold     = pvOutputThreshold,\n",
    "        useModel = useModel,\n",
    "        errorCovariance = errorCovariance,\n",
    "        verbose = TRUE,\n",
    "    #     output_file_name.cis = output_file_name_cis,\n",
    "    #     pvOutputThreshold.cis = pvOutputThreshold_cis,\n",
    "    #     snpspos = snpspos,\n",
    "    #     genepos = genepos,\n",
    "    #     cisDist = cisDist,\n",
    "        pvalue.hist = \"qqplot\",\n",
    "        min.pv.by.genesnp = FALSE,\n",
    "        noFDRsaveMemory = FALSE);\n",
    "    \n",
    "}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1ea73237",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "eeb31ce8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ac78df2b",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "62bbbeb8",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ce71dd99",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "53813a3c",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
