{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "1751dd1b",
   "metadata": {},
   "source": [
    "# Matrix eQTL"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "b7d1e6f0",
   "metadata": {},
   "outputs": [],
   "source": [
    "library(\"MatrixEQTL\");"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "faf50566",
   "metadata": {},
   "outputs": [],
   "source": [
    "base.dir = '/data_volume/memento/lupus/full_analysis/mateqtl/inputs/'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "80b1295c",
   "metadata": {},
   "outputs": [],
   "source": [
    "pop <- 'asian'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "211c7408",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Rows read: 100,000\n",
      "\n",
      "Rows read: 200,000\n",
      "\n",
      "Rows read: 300,000\n",
      "\n",
      "Rows read: 400,000\n",
      "\n",
      "Rows read: 500,000\n",
      "\n",
      "Rows read: 600,000\n",
      "\n",
      "Rows read: 700,000\n",
      "\n",
      "Rows read: 800,000\n",
      "\n",
      "Rows read: 900,000\n",
      "\n",
      "Rows read: 1,000,000\n",
      "\n",
      "Rows read: 1,100,000\n",
      "\n",
      "Rows read: 1,200,000\n",
      "\n",
      "Rows read: 1,300,000\n",
      "\n",
      "Rows read: 1,400,000\n",
      "\n",
      "Rows read: 1,500,000\n",
      "\n",
      "Rows read: 1,600,000\n",
      "\n",
      "Rows read: 1,700,000\n",
      "\n",
      "Rows read: 1,800,000\n",
      "\n",
      "Rows read: 1,900,000\n",
      "\n",
      "Rows read: 2,000,000\n",
      "\n",
      "Rows read: 2,100,000\n",
      "\n",
      "Rows read: 2,200,000\n",
      "\n",
      "Rows read: 2,300,000\n",
      "\n",
      "Rows read: 2,400,000\n",
      "\n",
      "Rows read: 2,500,000\n",
      "\n",
      "Rows read: 2,600,000\n",
      "\n",
      "Rows read: 2,700,000\n",
      "\n",
      "Rows read: 2,800,000\n",
      "\n",
      "Rows read: 2,900,000\n",
      "\n",
      "Rows read: 3,000,000\n",
      "\n",
      "Rows read: 3,100,000\n",
      "\n",
      "Rows read: 3,200,000\n",
      "\n",
      "Rows read: 3285470 done.\n",
      "\n",
      "Rows read: 35 done.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "useModel = modelLINEAR; # modelANOVA, modelLINEAR, or modelLINEAR_CROSS\n",
    "SNP_file_name = paste(base.dir, pop, '_genos.tsv', sep=\"\");\n",
    "snps_location_file_name = paste(base.dir, pop, \"_snpspos.txt\", sep=\"\");\n",
    "covariates_file_name = paste(base.dir,  pop, '_mateqtl_cov.txt', sep=\"\");\n",
    "pvOutputThreshold_cis = 1;\n",
    "errorCovariance = numeric();\n",
    "cisDist = 100000;\n",
    "\n",
    "gene_location_file_name = paste(base.dir, \"geneloc_tss_hg19.txt\", sep=\"\");\n",
    "\n",
    "\n",
    "## Load genotype data\n",
    "\n",
    "snps = SlicedData$new();\n",
    "snps$fileDelimiter = \"\\t\";      # the TAB character\n",
    "snps$fileOmitCharacters = \"NA\"; # denote missing values;\n",
    "snps$fileSkipRows = 1;          # one row of column labels\n",
    "snps$fileSkipColumns = 1;       # one column of row labels\n",
    "snps$fileSliceSize = 100000;      # read file in slices of 2,000 rows\n",
    "snps$LoadFile(SNP_file_name);\n",
    "\n",
    "\n",
    "## Load covariates\n",
    "\n",
    "cvrt = SlicedData$new();\n",
    "cvrt$fileDelimiter = \"\\t\";      # the TAB character\n",
    "cvrt$fileOmitCharacters = \"NA\"; # denote missing values;\n",
    "cvrt$fileSkipRows = 1;          # one row of column labels\n",
    "cvrt$fileSkipColumns = 1;       # one column of row labels\n",
    "if(length(covariates_file_name)>0) {\n",
    "cvrt$LoadFile(covariates_file_name);\n",
    "}\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "1b6fc457",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Rows read: 2,000\n",
      "\n",
      "Rows read: 3573 done.\n",
      "\n",
      "3521 of 3573 genes matched\n",
      "\n",
      "3285470 of 3285470 SNPs matched\n",
      "\n",
      "\n",
      " 1.51% done, 21,065 cis-eQTLs, 4,273 trans-eQTLs\n",
      "\n",
      " 3.03% done, 7,003 trans-eQTLs\n",
      "\n",
      " 4.54% done, 47,248 cis-eQTLs, 9,118 trans-eQTLs\n",
      "\n",
      " 6.06% done, 10,798 trans-eQTLs\n",
      "\n",
      " 7.57% done, 76,824 cis-eQTLs, 12,646 trans-eQTLs\n",
      "\n",
      " 9.09% done, 14,199 trans-eQTLs\n",
      "\n",
      "10.60% done, 96,169 cis-eQTLs, 16,112 trans-eQTLs\n",
      "\n",
      "12.12% done, 17,604 trans-eQTLs\n",
      "\n",
      "13.63% done, 141,483 cis-eQTLs, 19,075 trans-eQTLs\n",
      "\n",
      "15.15% done, 19,781 trans-eQTLs\n",
      "\n",
      "16.66% done, 163,207 cis-eQTLs, 21,624 trans-eQTLs\n",
      "\n",
      "18.18% done, 23,275 trans-eQTLs\n",
      "\n",
      "19.69% done, 174,127 cis-eQTLs, 24,743 trans-eQTLs\n",
      "\n",
      "21.21% done, 25,878 trans-eQTLs\n",
      "\n",
      "22.72% done, 187,569 cis-eQTLs, 28,395 trans-eQTLs\n",
      "\n",
      "24.24% done, 30,249 trans-eQTLs\n",
      "\n",
      "25.75% done, 208,235 cis-eQTLs, 31,931 trans-eQTLs\n",
      "\n",
      "27.27% done, 33,773 trans-eQTLs\n",
      "\n",
      "28.78% done, 227,425 cis-eQTLs, 35,930 trans-eQTLs\n",
      "\n",
      "30.30% done, 37,855 trans-eQTLs\n",
      "\n"
     ]
    }
   ],
   "source": [
    "snpspos = read.table(snps_location_file_name, header = TRUE, stringsAsFactors = FALSE);\n",
    "genepos = read.table(gene_location_file_name, header = TRUE, stringsAsFactors = FALSE, sep='\\t');\n",
    "\n",
    "for (ct in c('T8', 'T4', 'NK', 'ncM', 'cM', 'B')){\n",
    "# for (ct in c('B')){\n",
    "\n",
    "\n",
    "    expression_file_name = paste(base.dir, \"../pseudobulk/\", pop, '_', ct, '.csv',sep=\"\");\n",
    "    output_file_name = paste(base.dir, \"../outputs/\", pop, '_', ct, '_all_hg19.csv', sep=\"\");\n",
    "    ## Load gene expression data\n",
    "\n",
    "    gene = SlicedData$new();\n",
    "    gene$fileDelimiter = \"\\t\";      # the TAB character\n",
    "    gene$fileOmitCharacters = \"NA\"; # denote missing values;\n",
    "    gene$fileSkipRows = 1;          # one row of column labels\n",
    "    gene$fileSkipColumns = 1;       # one column of row labels\n",
    "    gene$fileSliceSize = 2000;      # read file in slices of 2,000 rows\n",
    "    gene$LoadFile(expression_file_name);\n",
    "\n",
    "\n",
    "    me = Matrix_eQTL_main(\n",
    "        snps = snps,\n",
    "        gene = gene,\n",
    "        cvrt = cvrt,\n",
    "#         output_file_name     = output_file_name,\n",
    "#         pvOutputThreshold     = pvOutputThreshold,\n",
    "        useModel = useModel,\n",
    "        errorCovariance = errorCovariance,\n",
    "        verbose = FALSE,\n",
    "        output_file_name.cis = output_file_name,\n",
    "        pvOutputThreshold.cis = pvOutputThreshold_cis,\n",
    "        snpspos = snpspos,\n",
    "        genepos = genepos,\n",
    "        cisDist = cisDist,\n",
    "        pvalue.hist = \"qqplot\",\n",
    "        min.pv.by.genesnp = FALSE,\n",
    "        noFDRsaveMemory = FALSE);\n",
    "    \n",
    "    }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "830059bf",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "R",
   "language": "R",
   "name": "ir"
  },
  "language_info": {
   "codemirror_mode": "r",
   "file_extension": ".r",
   "mimetype": "text/x-r-source",
   "name": "R",
   "pygments_lexer": "r",
   "version": "4.0.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
